---
title: "Arabidopsis Tandem Deamination Data Generation"
author: "Ben Morledge-Hampton"
date: "5/2/2022"
output: html_document
params:
  bioinformaticsDirectory:
    label: "Bioinformatics Projects Directory"
    value: C:/Bioinformatics_Projects
    input: select
    choices: [C:/Bioinformatics_Projects, F:/Bioinformatics_Projects]
  filtering:
    label: "Filtering"
    value: NONE
    input: select
    choices: [NONE, TGG_filtered]
  mismatchType:
    label: "Mismatch Type"
    value: CC>TT
    input: select
    choices: [CC>TT, GG>AA]
---


```{r setup_chunk}
library(knitr)
library(stringr)

expansionOffset = 10
cellType = "Arabidopsis"
CPDTimepoints = list(`ZT2` = NULL, `ZT14` = NULL)
minReadLength = 24; maxReadLength = 30

if (interactive()) {
  bioinformaticsDirectory = choose.dir(caption = "Select Bioinformatics_Projects Directory")
  filtering = readline("Filtering identifier (Input nothing for no filtering):")
  if (filtering == '') filtering = character()
  mismatchType = readline("Tandem mismatch type:")
} else {
  bioinformaticsDirectory = params$bioinformaticsDirectory
  filtering = params$filtering
  if (filtering == "NONE") filtering = character()
  mismatchType = params$mismatchType
}
if (is.na(bioinformaticsDirectory)) stop("No bioinformatics directory given.")

deaminationDataDirectory = file.path(bioinformaticsDirectory, "deamination_determination", "data", cellType)
mismatchesByReadDirectory = file.path(deaminationDataDirectory, "mismatches_by_read")
deaminationSequenceAndPositionDataDirectory = file.path(deaminationDataDirectory,"sequence_and_position_analysis")
deaminationExcisionSiteDataDirectory = file.path(deaminationDataDirectory, "excision_site_analysis")
dir.create(deaminationSequenceAndPositionDataDirectory, showWarnings = FALSE, recursive = TRUE)
dir.create(deaminationExcisionSiteDataDirectory, showWarnings = FALSE, recursive = TRUE)
```


```{r source_functions}
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "UsefulFileSystemFunctions.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationParseAndFilter.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationSequenceAndPositionAnalysis.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationAnchoredExcisionSiteAnalysis.R"))
```


```{r read_mismatch_data}
# Get file paths for the mismatch by read data
CPDMismatchData = list(rep1 = CPDTimepoints, rep2 = CPDTimepoints, all_reps = CPDTimepoints)
for (i in 1:length(CPDMismatchData)) {
  for (j in 1:length(CPDMismatchData[[i]])) {
    CPDMismatchData[[i]][[j]] = fread(generateFilePath(
      mismatchesByReadDirectory, dataTypeStrings$mismatchesByRead, ".bed",
      cellType = cellType, lesion = "CPD", timepoint = names(CPDMismatchData[[i]])[j],
      repitition = names(CPDMismatchData)[i], filtering = filtering,
      includedMismatches = mismatchType
      ))
  }
}
```


```{r generate_and_write_simplified_and_filtered_mismatch_data}
# Perform general filtering on the data
CPDMismatchData = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) filterResults(y, TRUE, 1, minReadLength, maxReadLength)
))

# Simplify the data
CPDMismatchData = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y)
))

# Write the simplified mismatch data
for (i in 1:length(CPDMismatchData)) {
  for (j in 1:length(CPDMismatchData[[i]])) {
    fwrite(CPDMismatchData[[i]][[j]], sep = '\t', na = "NA",
           file = generateFilePath(
             deaminationSequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
             cellType = cellType, lesion = "CPD", timepoint = names(CPDMismatchData[[i]])[j],
             repitition = names(CPDMismatchData)[i], filtering = filtering,
             includedMismatches = mismatchType
           ))
  }
}
```


```{r generate_and_write_mismatch_frequency_z-scores}
CPDMMFreqZScores = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) {
    getMismatchFrequencyZScoreByPosAndReadLength(y, 0)
  }
))

for (i in seq_along(CPDMMFreqZScores)) {
  for (j in seq_along(CPDMMFreqZScores[[i]])) {
    fwrite(CPDMMFreqZScores[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationSequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
      cellType = cellType, lesion = "CPD", repitition = names(CPDMMFreqZScores)[i],
      timepoint = names(CPDMMFreqZScores[[i]])[j], filtering = filtering, includedMismatches = c(mismatchType)
    ))
  }
}
```


```{r generate_and_write_read_length_counts}
CPDReadLengthCounts = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) {
    y[, .N, by = Read_Length][order(Read_Length)]
  }
))

for (i in seq_along(CPDReadLengthCounts)) {
  for (j in seq_along(CPDReadLengthCounts[[i]])) {
    fwrite(CPDReadLengthCounts[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationSequenceAndPositionDataDirectory, dataTypeStrings$readLengthcounts, ".tsv",
      cellType = cellType, lesion = "CPD", repitition = names(CPDReadLengthCounts)[i],
      timepoint = names(CPDReadLengthCounts[[i]])[j], filtering = filtering, includedMismatches = mismatchType
    ))
  }
}
```


```{r generate_and_write_mismatch_position_statistics_by_read_length_and_timepoint}
CPDGroupedStats = list(rep1 = list(), rep2 = list(), all_reps = list())
for (i in seq_along(CPDGroupedStats)) {
  CPDGroupedStats[[i]]$three_prime = getGroupedPositionStats(CPDMismatchData[[i]])
  CPDGroupedStats[[i]]$five_prime = getGroupedPositionStats(CPDMismatchData[[i]], posType = FIVE_PRIME)
}

for (i in seq_along(CPDGroupedStats)) {
  for (j in seq_along(CPDGroupedStats[[i]])) {
    fwrite(CPDGroupedStats[[i]][[j]], sep = '\t',
           file = generateFilePath(
             deaminationSequenceAndPositionDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
             cellType = cellType, lesion = "CPD", repitition = names(CPDGroupedStats)[i],
             strandPolarity = names(CPDGroupedStats[[i]])[j], filtering = filtering,
             includedMismatches = mismatchType
           ))
  }
}
```

EXCISION SITE ANALYSIS STARTS HERE

```{r read_expanded_mismatch_data}
# Get file paths for the expanded mismatch by read data
CPDMismatchData = list(rep1 = CPDTimepoints, rep2 = CPDTimepoints, all_reps = CPDTimepoints)
for (i in 1:length(CPDMismatchData)) {
  for (j in 1:length(CPDMismatchData[[i]])) {
    CPDMismatchData[[i]][[j]] = fread(generateFilePath(
      mismatchesByReadDirectory, dataTypeStrings$mismatchesByRead, ".bed",
      cellType = cellType, lesion = "CPD", timepoint = names(CPDMismatchData[[i]])[j],
      repitition = names(CPDMismatchData)[i], expansionNum = expansionOffset, filtering = filtering,
      includedMismatches = mismatchType
      ))
  }
}
```


```{r generate_simplified_and_filtered_mismatch_data}
# Filter the reads by length and mismatch type (single C>T).
CPDMismatchData = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) filterResults(y, minReadLength = minReadLength, maxReadLength = maxReadLength)
))

# Simplify the data
CPDMismatchData = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y, TRUE, expansionOffset = expansionOffset)
))

# Read mismatch frequency z-scores.
CPDMMFreqZScores = list(rep1 = CPDTimepoints, rep2 = CPDTimepoints, all_reps = CPDTimepoints)
for (i in seq_along(CPDMMFreqZScores)) {
  for (j in seq_along(CPDMMFreqZScores[[i]])) {
    CPDMMFreqZScores[[i]][[j]] = fread(generateFilePath(
      directory = deaminationSequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
      cellType = cellType, lesion = "CPD", repitition = names(CPDMMFreqZScores)[i],
      timepoint = names(CPDMMFreqZScores[[i]])[j], filtering = filtering, includedMismatches = mismatchType
    ))
  }
}

# Further filter data by read position based on length
# NOTE: 6-4 data uses CPD z-scores for filtering.
CPDMismatchData = mapply(function(CPDMismatchDataRep, CPDMMFreqZScoreRep) mapply(
  filterMismatchesByPositionAndReadLength, CPDMismatchDataRep, CPDMMFreqZScoreRep,
  expansionOffset = expansionOffset, SIMPLIFY = FALSE
), CPDMismatchData, CPDMMFreqZScores, SIMPLIFY = FALSE)
```


```{r orient_data_to_mismatch_positions}
# NOTE: The assignments to null are there so that the garbage collector can free up memory, and
#       I can still switch to new variable names for organizational purposes.
threePrimeMismatchAnchoredCPDData = lapply(CPDMismatchData, function(x) {
  lapply(x, orientDataToMismatch, THREE_PRIME, expansionOffset, TRUE)
})
fivePrimeMismatchAnchoredCPDData = lapply(CPDMismatchData, function(x) {
  lapply(x, orientDataToMismatch, FIVE_PRIME, expansionOffset, TRUE)
})
CPDMismatchData = NULL
```


```{r generate_and_write_nucleotide_frequency_data}
threePrimeMismatchCPDNucFreq = lapply(threePrimeMismatchAnchoredCPDData, tabulateNucFreqByPosTimepointAndLength)
for (i in seq_along(threePrimeMismatchCPDNucFreq)) {
  fwrite(threePrimeMismatchCPDNucFreq[[i]], sep = '\t', file = generateFilePath(
    directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
    cellType = cellType, lesion = "CPD", repitition = names(threePrimeMismatchCPDNucFreq)[i],
    strandPolarity = "three_prime", anchored = TRUE, filtering = filtering, includedMismatches = mismatchType
  ))
}

fivePrimeMismatchCPDNucFreq = lapply(fivePrimeMismatchAnchoredCPDData,
                                     tabulateNucFreqByPosTimepointAndLength, posType = FIVE_PRIME)
for (i in seq_along(fivePrimeMismatchCPDNucFreq)) {
  fwrite(fivePrimeMismatchCPDNucFreq[[i]], sep = '\t', file = generateFilePath(
    directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
    cellType = cellType, lesion = "CPD", repitition = names(fivePrimeMismatchCPDNucFreq)[i],
    strandPolarity = "five_prime", anchored = TRUE, filtering = filtering, includedMismatches = mismatchType
  ))
}
```


```{r generate_and_write_sequence_frequencies}
threePrimeMMAnchoredCPDSeqFreq = lapply(threePrimeMismatchAnchoredCPDData, function(mismatchDataRep) {
  lapply(mismatchDataRep, function(mismatchData) {
    getSequenceFreqByReadLengthAndPos(mismatchData, querySequences = c("CG","CA","TG","TA"))
  })
})
for (i in seq_along(threePrimeMMAnchoredCPDSeqFreq)) {
  for (j in seq_along(threePrimeMMAnchoredCPDSeqFreq[[i]])) {
    fwrite(threePrimeMMAnchoredCPDSeqFreq[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
      cellType = cellType, lesion = "CPD", repitition = names(threePrimeMMAnchoredCPDSeqFreq)[i],
      timepoint = names(threePrimeMMAnchoredCPDSeqFreq[[i]])[j], strandPolarity = "three_prime",
      anchored = TRUE, sequence = "YR", filtering = filtering, includedMismatches = mismatchType
    ))
  }
}

fivePrimeMMAnchoredCPDSeqFreq = lapply(fivePrimeMismatchAnchoredCPDData, function(mismatchDataRep) {
  lapply(mismatchDataRep, function(mismatchData) {
    getSequenceFreqByReadLengthAndPos(mismatchData, querySequences = c("TGG"), posType = FIVE_PRIME)
  })
})
for (i in seq_along(fivePrimeMMAnchoredCPDSeqFreq)) {
  for (j in seq_along(fivePrimeMMAnchoredCPDSeqFreq[[i]])) {
    fwrite(fivePrimeMMAnchoredCPDSeqFreq[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
      cellType = cellType, lesion = "CPD", repitition = names(fivePrimeMMAnchoredCPDSeqFreq)[i],
      timepoint = names(fivePrimeMMAnchoredCPDSeqFreq[[i]])[j], strandPolarity = "five_prime",
      anchored = TRUE, sequence = "TGG", filtering = filtering, includedMismatches = mismatchType
    ))
  }
}
```


