---
title: "GM12878 Deamination Anchored Excision Site Data Generation"
author: "Ben Morledge-Hampton"
date: "5/2/2022"
output: html_document
params:
  bioinformaticsDirectory:
    label: "Bioinformatics Projects Directory"
    value: C:/Bioinformatics_Projects
    input: select
    choices: [C:/Bioinformatics_Projects, F:/Bioinformatics_Projects]
  filtering:
    label: "Filtering"
    value: NONE
    input: select
    choices: [NONE, TGG_filtered]
---


```{r setup_chunk}
library(knitr)

expansionOffset = 10
cellType = "GM12878"
CisplatinTimepoints = list(`90min` = NULL)
minReadLength = 22; maxReadLength = 30

if (interactive()) {
  bioinformaticsDirectory = choose.dir(caption = "Select Bioinformatics_Projects Directory")
  filtering = readline("Filtering identifier (Input nothing for no filtering):")
  if (filtering == '') filtering = character()
} else {
  bioinformaticsDirectory = params$bioinformaticsDirectory
  filtering = params$filtering
  if (filtering == "NONE") filtering = character()
}
if (is.na(bioinformaticsDirectory)) stop("No bioinformatics directory given.")

deaminationDataDirectory = file.path(bioinformaticsDirectory, "deamination_determination", "data", cellType)
NHF1DeaminationDataDirectory = file.path(bioinformaticsDirectory, "deamination_determination", "data", "NHF1")
mismatchesByReadDirectory = file.path(deaminationDataDirectory, "mismatches_by_read")
deaminationSequenceAndPositionDataDirectory = file.path(deaminationDataDirectory,"sequence_and_position_analysis")
NHF1DeaminationSequenceAndPositionDataDirectory = file.path(NHF1DeaminationDataDirectory,"sequence_and_position_analysis")
deaminationExcisionSiteDataDirectory = file.path(deaminationDataDirectory, "excision_site_analysis")
dir.create(deaminationExcisionSiteDataDirectory, showWarnings = FALSE, recursive = TRUE)
```


```{r source_functions}
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "UsefulFileSystemFunctions.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationParseAndFilter.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationAnchoredExcisionSiteAnalysis.R"))
```


```{r read_expanded_mismatch_data}
# Get file paths for the expanded mismatch by read data
CisplatinMismatchData = list(rep1 = CisplatinTimepoints,
                             rep2 = CisplatinTimepoints)
for (i in 1:length(CisplatinMismatchData)) {
  for (j in 1:length(CisplatinMismatchData[[i]])) {
    CisplatinMismatchData[[i]][[j]] = fread(generateFilePath(
      mismatchesByReadDirectory, dataTypeStrings$mismatchesByRead, ".bed",
      cellType = cellType, lesion = "cisplatin", timepoint = names(CisplatinMismatchData[[i]])[j],
      repitition = names(CisplatinMismatchData)[i], expansionNum = expansionOffset, filtering = filtering
      ))
  }
}
```


```{r generate_simplified_and_filtered_mismatch_data}
# Filter the reads by length and mismatch type (single C>T).
CisplatinMismatchData = lapply(CisplatinMismatchData, function(x) lapply(
  x, function(y) filterResults(y[V5 == "C>T"], minReadLength = minReadLength, maxReadLength = maxReadLength)
))

# Simplify the data
CisplatinMismatchData = lapply(CisplatinMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y, TRUE, expansionOffset = expansionOffset)
))

# Read mismatch frequency z-scores.
# NOTE: Using NHF1 CPD scores. Otherwise, data will be too tightly constrained.
NHF1CPDMMFreqZScores = fread(generateFilePath(
      directory = NHF1DeaminationSequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
      cellType = "NHF1", lesion = "CPD", repitition = "rep1",
      timepoint = "1h", filtering = filtering, includedMismatches = c("C>T")
))


# Further filter data by read position based on length
CisplatinMismatchData = lapply(CisplatinMismatchData, function(CisplatinMismatchDataRep) {
  lapply(CisplatinMismatchDataRep, filterMismatchesByPositionAndReadLength,
         NHF1CPDMMFreqZScores, expansionOffset = expansionOffset)
})
```


```{r orient_data_to_mismatch_positions}
# NOTE: The assignment to null is there so that the garbage collector can free up memory, and
#       I can still switch to new variable names for organizational purposes.
threePrimeMismatchAnchoredCisplatinData = lapply(CisplatinMismatchData, function(x) {
  lapply(x, orientDataToMismatch, THREE_PRIME, expansionOffset)
})
fivePrimeMismatchAnchoredCisplatinData = lapply(CisplatinMismatchData, function(x) {
  lapply(x, orientDataToMismatch, FIVE_PRIME, expansionOffset)
})
CisplatinMismatchData = NULL
```


```{r generate_and_write_nucleotide_frequency_data}
threePrimeMismatchCisplatinNucFreq = lapply(threePrimeMismatchAnchoredCisplatinData, tabulateNucFreqByPosTimepointAndLength)
for (i in seq_along(threePrimeMismatchCisplatinNucFreq)) {
  fwrite(threePrimeMismatchCisplatinNucFreq[[i]], sep = '\t', file = generateFilePath(
    directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
    cellType = cellType, lesion = "cisplatin", repitition = names(threePrimeMismatchCisplatinNucFreq)[i],
    strandPolarity = "three_prime", anchored = TRUE, filtering = filtering
  ))
}

fivePrimeMismatchCisplatinNucFreq = lapply(fivePrimeMismatchAnchoredCisplatinData,
                                           tabulateNucFreqByPosTimepointAndLength, posType = FIVE_PRIME)
for (i in seq_along(fivePrimeMismatchCisplatinNucFreq)) {
  fwrite(fivePrimeMismatchCisplatinNucFreq[[i]], sep = '\t', file = generateFilePath(
    directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
    cellType = cellType, lesion = "cisplatin", repitition = names(fivePrimeMismatchCisplatinNucFreq)[i],
    strandPolarity = "five_prime", anchored = TRUE, filtering = filtering
  ))
}
```


```{r generate_and_write_sequence_enrichment}
threePrimeMMAnchoredCisplatinEnrichment = mapply(function(mismatchDataRep, nucFreqRep) {
  setNames(lapply(seq_along(mismatchDataRep), function(i) {
    getSequenceFreqByReadLengthAndPos(mismatchDataRep[[i]],
                                      nucFreqRep[Timepoint == names(mismatchDataRep)[i]],
                                      c("CG","CA","TG","TA"))
  }), names(mismatchDataRep))
}, threePrimeMismatchAnchoredCisplatinData, threePrimeMismatchCisplatinNucFreq, SIMPLIFY = FALSE)
for (i in seq_along(threePrimeMMAnchoredCisplatinEnrichment)) {
  for (j in seq_along(threePrimeMMAnchoredCisplatinEnrichment[[i]])) {
    fwrite(threePrimeMMAnchoredCisplatinEnrichment[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$sequenceEnrichment, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(threePrimeMMAnchoredCisplatinEnrichment)[i],
      timepoint = names(threePrimeMMAnchoredCisplatinEnrichment[[i]])[j], strandPolarity = "three_prime",
      anchored = TRUE, sequence = "YR", filtering = filtering
    ))
  }
}

fivePrimeMMAnchoredCisplatinEnrichment = mapply(function(mismatchDataRep, nucFreqRep) {
  setNames(lapply(seq_along(mismatchDataRep), function(i) {
    getSequenceFreqByReadLengthAndPos(mismatchDataRep[[i]],
                                      nucFreqRep[Timepoint == names(mismatchDataRep)[i]],
                                      c("TGG"), posType = FIVE_PRIME)
  }), names(mismatchDataRep))
}, fivePrimeMismatchAnchoredCisplatinData, fivePrimeMismatchCisplatinNucFreq, SIMPLIFY = FALSE)
for (i in seq_along(fivePrimeMMAnchoredCisplatinEnrichment)) {
  for (j in seq_along(fivePrimeMMAnchoredCisplatinEnrichment[[i]])) {
    fwrite(fivePrimeMMAnchoredCisplatinEnrichment[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$sequenceEnrichment, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(fivePrimeMMAnchoredCisplatinEnrichment)[i],
      timepoint = names(fivePrimeMMAnchoredCisplatinEnrichment[[i]])[j], strandPolarity = "five_prime",
      anchored = TRUE, sequence = "TGG", filtering = filtering
    ))
  }
}
```


```{r generate_and_write_sequence_frequencies}
threePrimeMMAnchoredCisplatinSeqFreq = lapply(threePrimeMismatchAnchoredCisplatinData, function(mismatchDataRep) {
  lapply(mismatchDataRep, function(mismatchData) {
    getSequenceFreqByReadLengthAndPos(mismatchData, querySequences = c("CG","CA","TG","TA"))
  })
})
for (i in seq_along(threePrimeMMAnchoredCisplatinSeqFreq)) {
  for (j in seq_along(threePrimeMMAnchoredCisplatinSeqFreq[[i]])) {
    fwrite(threePrimeMMAnchoredCisplatinSeqFreq[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(threePrimeMMAnchoredCisplatinSeqFreq)[i],
      timepoint = names(threePrimeMMAnchoredCisplatinSeqFreq[[i]])[j], strandPolarity = "three_prime",
      anchored = TRUE, sequence = "YR", filtering = filtering
    ))
  }
}

fivePrimeMMAnchoredCisplatinSeqFreq = lapply(fivePrimeMismatchAnchoredCisplatinData, function(mismatchDataRep) {
  lapply(mismatchDataRep, function(mismatchData) {
    getSequenceFreqByReadLengthAndPos(mismatchData, querySequences = c("TGG"), posType = FIVE_PRIME)
  })
})
for (i in seq_along(fivePrimeMMAnchoredCisplatinSeqFreq)) {
  for (j in seq_along(fivePrimeMMAnchoredCisplatinSeqFreq[[i]])) {
    fwrite(fivePrimeMMAnchoredCisplatinSeqFreq[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = deaminationExcisionSiteDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(fivePrimeMMAnchoredCisplatinSeqFreq)[i],
      timepoint = names(fivePrimeMMAnchoredCisplatinSeqFreq[[i]])[j], strandPolarity = "five_prime",
      anchored = TRUE, sequence = "TGG", filtering = filtering
    ))
  }
}
```

