---
title: "NHF1 Tandem Deamination Data Generation"
author: "Ben Morledge-Hampton"
date: "5/2/2022"
output: html_document
params:
  bioinformaticsDirectory:
    label: "Bioinformatics Projects Directory"
    value: C:/Bioinformatics_Projects
    input: select
    choices: [C:/Bioinformatics_Projects, F:/Bioinformatics_Projects]
  filtering:
    label: "Filtering"
    value: NONE
    input: select
    choices: [NONE, TGG_filtered]
  mismatchType:
    label: "Mismatch Type"
    value: CC>TT
    input: select
    choices: [CC>TT, GG>AA]
---


```{r setup_chunk}
library(knitr)
library(stringr)

expansionOffset = 10
cellType = "GM12878"
cisplatinTimepoints = list(`90min` = NULL)
minReadLength = 22; maxReadLength = 30

if (interactive()) {
  bioinformaticsDirectory = choose.dir(caption = "Select Bioinformatics_Projects Directory")
  filtering = readline("Filtering identifier (Input nothing for no filtering):")
  if (filtering == '') filtering = character()
  mismatchType = readline("Tandem mismatch type:")
} else {
  bioinformaticsDirectory = params$bioinformaticsDirectory
  filtering = params$filtering
  if (filtering == "NONE") filtering = character()
  mismatchType = params$mismatchType
}
if (is.na(bioinformaticsDirectory)) stop("No bioinformatics directory given.")

deaminationDataDirectory = file.path(bioinformaticsDirectory, "deamination_determination", "data", cellType)
tandemDeaminationDataDirectory = file.path(deaminationDataDirectory,
                                           paste(str_replace(mismatchType, '>', "_to_"),
                                                 "tandem_deaminations", sep = '_'))
dir.create(tandemDeaminationDataDirectory, recursive = TRUE, showWarnings = FALSE)
```


```{r source_functions}
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "UsefulFileSystemFunctions.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationParseAndFilter.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationSequenceAndPositionAnalysis.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationAnchoredExcisionSiteAnalysis.R"))
```


```{r read_mismatch_data}
# Get file paths for the mismatch by read data
cisplatinMismatchData = list(rep1 = cisplatinTimepoints,
                             rep2 = cisplatinTimepoints)
for (i in 1:length(cisplatinMismatchData)) {
  for (j in 1:length(cisplatinMismatchData[[i]])) {
    cisplatinMismatchData[[i]][[j]] = fread(generateFilePath(
      tandemDeaminationDataDirectory, dataTypeStrings$mismatchesByRead, ".bed",
      cellType = cellType, lesion = "cisplatin", timepoint = names(cisplatinMismatchData[[i]])[j],
      repitition = names(cisplatinMismatchData)[i], filtering = filtering,
      includedMismatches = mismatchType
      ))
  }
}
```


```{r generate_and_write_simplified_and_filtered_mismatch_data}
# Perform general filtering on the data
cisplatinMismatchData = lapply(cisplatinMismatchData, function(x) lapply(
  x, function(y) filterResults(y, TRUE, 1, minReadLength, maxReadLength)
))

# Simplify the data
cisplatinMismatchData = lapply(cisplatinMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y)
))

# Write the simplified mismatch data
for (i in 1:length(cisplatinMismatchData)) {
  for (j in 1:length(cisplatinMismatchData[[i]])) {
    fwrite(cisplatinMismatchData[[i]][[j]], sep = '\t', na = "NA",
           file = generateFilePath(
             tandemDeaminationDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
             cellType = cellType, lesion = "cisplatin", timepoint = names(cisplatinMismatchData[[i]])[j],
             repitition = names(cisplatinMismatchData)[i], filtering = filtering,
             includedMismatches = mismatchType
           ))
  }
}
```


```{r generate_and_write_mismatch_frequency_z-scores}
cisplatinMMFreqZScores = lapply(cisplatinMismatchData, function(x) lapply(
  x, function(y) {
    getMismatchFrequencyZScoreByPosAndReadLength(y, 0)
  }
))

for (i in seq_along(cisplatinMMFreqZScores)) {
  for (j in seq_along(cisplatinMMFreqZScores[[i]])) {
    fwrite(cisplatinMMFreqZScores[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = tandemDeaminationDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(cisplatinMMFreqZScores)[i],
      timepoint = names(cisplatinMMFreqZScores[[i]])[j], filtering = filtering, includedMismatches = c(mismatchType)
    ))
  }
}
```


```{r generate_and_write_mismatch_position_statistics_by_read_length_and_timepoint}
cisplatinGroupedStats = list(rep1 = list(), rep2 = list())
cisplatinGroupedStats$rep1$three_prime = getGroupedPositionStats(cisplatinMismatchData$rep1)
cisplatinGroupedStats$rep2$three_prime = getGroupedPositionStats(cisplatinMismatchData$rep2)
cisplatinGroupedStats$rep1$five_prime = getGroupedPositionStats(cisplatinMismatchData$rep1, posType = FIVE_PRIME)
cisplatinGroupedStats$rep2$five_prime = getGroupedPositionStats(cisplatinMismatchData$rep2, posType = FIVE_PRIME)

for (i in seq_along(cisplatinGroupedStats)) {
  for (j in seq_along(cisplatinGroupedStats[[i]])) {
    fwrite(cisplatinGroupedStats[[i]][[j]], sep = '\t',
           file = generateFilePath(
             tandemDeaminationDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
             cellType = cellType, lesion = "cisplatin", repitition = names(cisplatinGroupedStats)[i],
             strandPolarity = names(cisplatinGroupedStats[[i]])[j], filtering = filtering,
             includedMismatches = mismatchType
           ))
  }
}
```

EXCISION SITE ANALYSIS STARTS HERE

```{r read_expanded_mismatch_data}
# Get file paths for the expanded mismatch by read data
cisplatinMismatchData = list(rep1 = cisplatinTimepoints,
                             rep2 = cisplatinTimepoints)
for (i in 1:length(cisplatinMismatchData)) {
  for (j in 1:length(cisplatinMismatchData[[i]])) {
    cisplatinMismatchData[[i]][[j]] = fread(generateFilePath(
      tandemDeaminationDataDirectory, dataTypeStrings$mismatchesByRead, ".bed",
      cellType = cellType, lesion = "cisplatin", timepoint = names(cisplatinMismatchData[[i]])[j],
      repitition = names(cisplatinMismatchData)[i], expansionNum = expansionOffset, filtering = filtering,
      includedMismatches = mismatchType
      ))
  }
}
```


```{r generate_simplified_and_filtered_mismatch_data}
# Filter the reads by length and mismatch type (single C>T).
cisplatinMismatchData = lapply(cisplatinMismatchData, function(x) lapply(
  x, function(y) filterResults(y, minReadLength = minReadLength, maxReadLength = maxReadLength)
))

# Simplify the data
cisplatinMismatchData = lapply(cisplatinMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y, TRUE, expansionOffset = expansionOffset)
))

# Read mismatch frequency z-scores.
cisplatinMMFreqZScores = list(rep1 = cisplatinTimepoints, rep2 = cisplatinTimepoints)
for (i in seq_along(cisplatinMMFreqZScores)) {
  for (j in seq_along(cisplatinMMFreqZScores[[i]])) {
    cisplatinMMFreqZScores[[i]][[j]] = fread(generateFilePath(
      directory = tandemDeaminationDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(cisplatinMMFreqZScores)[i],
      timepoint = names(cisplatinMMFreqZScores[[i]])[j], filtering = filtering, includedMismatches = mismatchType
    ))
  }
}

# Further filter data by read position based on length
cisplatinMismatchData = mapply(function(cisplatinMismatchDataRep, cisplatinMMFreqZScoreRep) mapply(
  filterMismatchesByPositionAndReadLength, cisplatinMismatchDataRep, cisplatinMMFreqZScoreRep,
  expansionOffset = expansionOffset, SIMPLIFY = FALSE
), cisplatinMismatchData, cisplatinMMFreqZScores, SIMPLIFY = FALSE)
```


```{r orient_data_to_mismatch_positions}
# NOTE: The assignments to null are there so that the garbage collector can free up memory, and
#       I can still switch to new variable names for organizational purposes.
threePrimeMismatchAnchoredCisplatinData = lapply(cisplatinMismatchData, function(x) {
  lapply(x, orientDataToMismatch, THREE_PRIME, expansionOffset, TRUE)
})
fivePrimeMismatchAnchoredCisplatinData = lapply(cisplatinMismatchData, function(x) {
  lapply(x, orientDataToMismatch, FIVE_PRIME, expansionOffset, TRUE)
})
cisplatinMismatchData = NULL
```


```{r generate_and_write_nucleotide_frequency_data}
threePrimeMismatchCisplatinNucFreq = lapply(threePrimeMismatchAnchoredCisplatinData,
                                            tabulateNucFreqByPosTimepointAndLength)
for (i in seq_along(threePrimeMismatchCisplatinNucFreq)) {
  fwrite(threePrimeMismatchCisplatinNucFreq[[i]], sep = '\t', file = generateFilePath(
    directory = tandemDeaminationDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
    cellType = cellType, lesion = "cisplatin", repitition = names(threePrimeMismatchCisplatinNucFreq)[i],
    strandPolarity = "three_prime", anchored = TRUE, filtering = filtering, includedMismatches = mismatchType
  ))
}

fivePrimeMismatchCisplatinNucFreq = lapply(fivePrimeMismatchAnchoredCisplatinData,
                                           tabulateNucFreqByPosTimepointAndLength, posType = FIVE_PRIME)
for (i in seq_along(fivePrimeMismatchCisplatinNucFreq)) {
  fwrite(fivePrimeMismatchCisplatinNucFreq[[i]], sep = '\t', file = generateFilePath(
    directory = tandemDeaminationDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
    cellType = cellType, lesion = "cisplatin", repitition = names(fivePrimeMismatchCisplatinNucFreq)[i],
    strandPolarity = "five_prime", anchored = TRUE, filtering = filtering, includedMismatches = mismatchType
  ))
}
```


```{r generate_and_write_sequence_frequencies}
threePrimeMMAnchoredCisplatinSeqFreq = lapply(
  threePrimeMismatchAnchoredCisplatinData,function(mismatchDataRep) {
    lapply(mismatchDataRep, function(mismatchData) {
      getSequenceFreqByReadLengthAndPos(mismatchData, querySequences = c("CG","CA","TG","TA"))
    })
  }
)
for (i in seq_along(threePrimeMMAnchoredCisplatinSeqFreq)) {
  for (j in seq_along(threePrimeMMAnchoredCisplatinSeqFreq[[i]])) {
    fwrite(threePrimeMMAnchoredCisplatinSeqFreq[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = tandemDeaminationDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(threePrimeMMAnchoredCisplatinSeqFreq)[i],
      timepoint = names(threePrimeMMAnchoredCisplatinSeqFreq[[i]])[j], strandPolarity = "three_prime",
      anchored = TRUE, sequence = "YR", filtering = filtering, includedMismatches = mismatchType
    ))
  }
}

fivePrimeMMAnchoredCisplatinSeqFreq = lapply(
  fivePrimeMismatchAnchoredCisplatinData, function(mismatchDataRep) {
    lapply(mismatchDataRep, function(mismatchData) {
      getSequenceFreqByReadLengthAndPos(mismatchData, querySequences = c("TGG"), posType = FIVE_PRIME)
    })
  }
)
for (i in seq_along(fivePrimeMMAnchoredCisplatinSeqFreq)) {
  for (j in seq_along(fivePrimeMMAnchoredCisplatinSeqFreq[[i]])) {
    fwrite(fivePrimeMMAnchoredCisplatinSeqFreq[[i]][[j]], sep = '\t', file = generateFilePath(
      directory = tandemDeaminationDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
      cellType = cellType, lesion = "cisplatin", repitition = names(fivePrimeMMAnchoredCisplatinSeqFreq)[i],
      timepoint = names(fivePrimeMMAnchoredCisplatinSeqFreq[[i]])[j], strandPolarity = "five_prime",
      anchored = TRUE, sequence = "TGG", filtering = filtering, includedMismatches = mismatchType
    ))
  }
}
```


