---
title: "R Notebook"
output: html_document
params:
  bioinformaticsDirectory:
    label: "Bioinformatics Projects Directory"
    value: C:/Bioinformatics_Projects
    input: select
    choices: [C:/Bioinformatics_Projects, F:/Bioinformatics_Projects]
---


```{r setup_chunk}
library(knitr)
opts_chunk$set(dev = "png", dpi = 600, fig.width = 7, fig.height = 5, 
               fig.path = paste0(file.path("deamination_analysis_output","human_analysis_figures"),
                                 .Platform$file.sep))
if (interactive()) {
  bioinformaticsDirectory = choose.dir(caption = "Select Bioinformatics_Projects Directory")
} else {
  bioinformaticsDirectory = params$bioinformaticsDirectory
}

deaminationDataDirectory = file.path(bioinformaticsDirectory, "deamination_determination", "data", "human")
```


```{r source_functions}
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R', "DeaminationAnalysis.R"))
source(file.path(bioinformaticsDirectory, "Chromatin_Features_Analysis", "R_scripts", "PeriodicityAnalysis.R"))
```



```{r acquire_and_format_data_1}
# Read in the mismatch by read data
CPDMismatchDataFilePaths = list(rep1 = list(), rep2 = list())
CPDMismatchDataFilePaths$rep1 = list(
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_1h_Rep1_mismatches_by_read.bed"),
  hr8 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_8h_Rep1_mismatches_by_read.bed"),
  hr24 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_24h_Rep1_mismatches_by_read.bed"),
  hr48 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_48h_Rep1_mismatches_by_read.bed")
)
CPDMismatchDataFilePaths$rep2 = list(
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_1h_Rep2_mismatches_by_read.bed"),
  hr8 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_8h_Rep2_mismatches_by_read.bed"),
  hr24 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_24h_Rep2_mismatches_by_read.bed"),
  hr48 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_CPD_48h_Rep2_mismatches_by_read.bed")
)

sixFourMismatchDataFilePaths = list(rep1 = list(), rep2 = list())
sixFourMismatchDataFilePaths$rep1 = list(
  min5 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_6-4-PP_5min_Rep1_mismatches_by_read.bed"),
  min20 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_6-4-PP_20min_Rep1_mismatches_by_read.bed"),
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_6-4-PP_1h_Rep1_mismatches_by_read.bed")
)
sixFourMismatchDataFilePaths$rep2 = list(
  min5 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_6-4-PP_5min_Rep2_mismatches_by_read.bed"),
  min20 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_6-4-PP_20min_Rep2_mismatches_by_read.bed"),
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read", "NHF1_6-4-PP_1h_Rep2_mismatches_by_read.bed")
)

CPDMismatchData = lapply(CPDMismatchDataFilePaths, function(x) lapply(x, fread))
sixFourMismatchData = lapply(sixFourMismatchDataFilePaths, function(x) lapply(x, fread))

# Create a subset of single mismatch C>T reads.
CPDSingleCToTMismatchData = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) filterResults(y[V5 == "C>T"], minReadLength = 23, maxReadLength = 31)
))
sixFourSingleCToTMismatchData = lapply(sixFourMismatchData, function(x) lapply(
  x, function(y) filterResults(y[V5 == "C>T"], minReadLength = 23, maxReadLength = 31)
))

# Perform general filtering on the data
CPDMismatchData = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) filterResults(y, TRUE, 3, 23, 31)
))
sixFourMismatchData = lapply(sixFourMismatchData, function(x) lapply(
  x, function(y) filterResults(y, TRUE, 3, 23, 31)
))

# # Get the tables of read length vs. GC content
# CPDReadLengthVsGCData = lapply(CPDMismatchData, function(x) lapply(
#   x, getReadLengthAndGCContent
# ))
# sixReadLengthVsGCData = lapply(sixFourMismatchData, function(x) lapply(
#   x, getReadLengthAndGCContent
# ))

# # Extract and simplify tandem C>T mismatches
# CPDTandemCTMismatches = lapply(CPDMismatchData, function(x) lapply(
#   x, function(y) simplifyTable(getTandemCTMismatches(y))
# ))
# sixFourTandemCTMismatches = lapply(sixFourMismatchData, function(x) lapply(
#   x, function(y) simplifyTable(getTandemCTMismatches(y))
# ))

# Simplify the data
CPDMismatchData = lapply(CPDMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y)
))
sixFourMismatchData = lapply(sixFourMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y)
))
CPDSingleCToTMismatchData = lapply(CPDSingleCToTMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y, TRUE)
))
sixFourSingleCToTMismatchData = lapply(sixFourSingleCToTMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y, TRUE)
))

```


```{r CPD_mismatch_type_frequency}
for (i in 1:length(CPDMismatchData)) {
  for (j in 1:length(CPDMismatchData[[i]])) {
    plotMismatchTypeFrequencies(CPDMismatchData[[i]][[j]],
                                title = paste("CPD", names(CPDMismatchData)[i], 
                                              names(CPDMismatchData[[i]])[j], "Mismatch Type Frequency"))
  }
}
```


```{r 6-4_mismatch_type_frequency}
for (i in 1:length(sixFourMismatchData)) {
  for (j in 1:length(sixFourMismatchData[[i]])) {
    plotMismatchTypeFrequencies(sixFourMismatchData[[i]][[j]],
                                title = paste("6-4", names(sixFourMismatchData)[i], 
                                              names(sixFourMismatchData[[i]])[j], "Mismatch Type Frequency"))
  }
}

```


```{r CPD_mismatch_pos_frequency}
for (i in 1:length(CPDMismatchData)) {
  for (j in 1:length(CPDMismatchData[[i]])) {
    plotMismatchPositionFrequencies(CPDMismatchData[[i]][[j]], includedTypes = c("C>T"),
                                    title = paste("CPD", names(CPDMismatchData)[i], 
                                                  names(CPDMismatchData[[i]])[j], "C>T Mismatch Positions"),
                                    posType = FIVE_PRIME)
    plotMismatchPositionFrequencies(CPDMismatchData[[i]][[j]], omittedTypes = c("C>T"),
                                    title = paste("CPD", names(CPDMismatchData)[i], 
                                                  names(CPDMismatchData[[i]])[j], "!C>T Mismatch Positions"))
  }
}
```


```{r 6-4_mismatch_pos_frequency}
for (i in 1:length(sixFourMismatchData)) {
  for (j in 1:length(sixFourMismatchData[[i]])) {
    plotMismatchPositionFrequencies(sixFourMismatchData[[i]][[j]], includedTypes = c("C>T"),
                                    title = paste("6-4", names(sixFourMismatchData)[i], 
                                                  names(sixFourMismatchData[[i]])[j], "C>T Mismatch Positions"))
    plotMismatchPositionFrequencies(sixFourMismatchData[[i]][[j]], includedTypes = c("C>T"),
                                    title = paste("6-4", names(sixFourMismatchData)[i], 
                                                  names(sixFourMismatchData[[i]])[j], "C>T Mismatch Positions"),
                                    posType = FIVE_PRIME)
    
    plotMismatchPositionFrequencies(sixFourMismatchData[[i]][[j]], omittedTypes = c("C>T"),
                                    title = paste("6-4", names(sixFourMismatchData)[i], 
                                                  names(sixFourMismatchData[[i]])[j], "!C>T Mismatch Positions"))
  }
}

```


# ```{r CPD_tandem_C_to_T_pos_frequency}
# for (i in 1:length(CPDTandemCTMismatches)) {
#   for (j in 1:length(CPDTandemCTMismatches[[i]])) {
#     plotMismatchPositionFrequencies(CPDTandemCTMismatches[[i]][[j]],
#                                     title = paste("CPD", names(CPDTandemCTMismatches)[i], 
#                                                   names(CPDTandemCTMismatches[[i]])[j], "Tandem C>T Mismatch Pos"))
#   }
# }
# ```


# ```{r 6-4_tandem_C_to_T_pos_frequency}
# for (i in 1:length(sixFourTandemCTMismatches)) {
#   for (j in 1:length(sixFourTandemCTMismatches[[i]])) {
#     plotMismatchPositionFrequencies(sixFourTandemCTMismatches[[i]][[j]],
#                                     title = paste("6-4", names(sixFourTandemCTMismatches)[i], 
#                                                   names(sixFourTandemCTMismatches[[i]])[j], 
#                                                   "Tandem C>T Mismatch Pos"))
#   }
# }
# ```


```{r facet_mismatch_position_frequencies, fig.width = 9, fig.height = 5}
plotPositionAcrossTimepointAndReadLength(CPDMismatchData$rep1, includedTypes = c("C>T"),
                                         title = "CPD C>T Mismatch Pos Freq Rep1")
plotPositionAcrossTimepointAndReadLength(CPDMismatchData$rep2, includedTypes = c("C>T"),
                                         title = "CPD C>T Mismatch Pos Freq Rep2")
plotPositionAcrossTimepointAndReadLength(CPDMismatchData$rep1, includedTypes = c("C>T"),
                                         title = "CPD C>T Mismatch Pos Freq Rep1", posType = FIVE_PRIME)
plotPositionAcrossTimepointAndReadLength(CPDMismatchData$rep2, includedTypes = c("C>T"),
                                         title = "CPD C>T Mismatch Pos Freq Rep2", posType = FIVE_PRIME)

plotPositionAcrossTimepointAndReadLength(sixFourMismatchData$rep1, includedTypes = c("C>T"),
                                         title = "6-4 C>T Mismatch Pos Freq Rep1")
plotPositionAcrossTimepointAndReadLength(sixFourMismatchData$rep2, includedTypes = c("C>T"),
                                         title = "6-4 C>T Mismatch Pos Freq Rep2")
plotPositionAcrossTimepointAndReadLength(sixFourMismatchData$rep1, includedTypes = c("C>T"),
                                         title = "6-4 C>T Mismatch Pos Freq Rep1", posType = FIVE_PRIME)
plotPositionAcrossTimepointAndReadLength(sixFourMismatchData$rep2, includedTypes = c("C>T"),
                                         title = "6-4 C>T Mismatch Pos Freq Rep2", posType = FIVE_PRIME)
```


```{r facet_mismatch_position_statistics}
CPDGroupedStats = list(rep1 = list(), rep2 = list())
CPDGroupedStats$rep1$three_prime = getGroupedPositionStats(CPDMismatchData$rep1, includedTypes = c("C>T"))
CPDGroupedStats$rep2$three_prime = getGroupedPositionStats(CPDMismatchData$rep2, includedTypes = c("C>T"))
CPDGroupedStats$rep1$five_prime =
  getGroupedPositionStats(CPDMismatchData$rep1, includedTypes = c("C>T"), posType = FIVE_PRIME)
CPDGroupedStats$rep2$five_prime =
  getGroupedPositionStats(CPDMismatchData$rep2, includedTypes = c("C>T"), posType = FIVE_PRIME)

sixFourGroupedStats = list(rep1 = list(), rep2 = list())
sixFourGroupedStats$rep1$three_prime = getGroupedPositionStats(sixFourMismatchData$rep1, includedTypes = c("C>T"))
sixFourGroupedStats$rep2$three_prime = getGroupedPositionStats(sixFourMismatchData$rep2, includedTypes = c("C>T"))
sixFourGroupedStats$rep1$five_prime =
  getGroupedPositionStats(sixFourMismatchData$rep1, includedTypes = c("C>T"), posType = FIVE_PRIME)
sixFourGroupedStats$rep2$five_prime =
  getGroupedPositionStats(sixFourMismatchData$rep2, includedTypes = c("C>T"), posType = FIVE_PRIME)

tableOutputDir = file.path("deamination_analysis_output","human_analysis_tables")
dir.create(tableOutputDir, recursive = TRUE)

fwrite(CPDGroupedStats$rep1$three_prime,
       file.path(tableOutputDir, "CPD_mismatch_three_prime_position_stats_rep1.tsv"),sep = '\t')
fwrite(CPDGroupedStats$rep2$three_prime,
       file.path(tableOutputDir, "CPD_mismatch_three_prime_position_stats_rep2.tsv"),sep = '\t')
fwrite(CPDGroupedStats$rep1$five_prime,
       file.path(tableOutputDir, "CPD_mismatch_five_prime_position_stats_rep1.tsv"),sep = '\t')
fwrite(CPDGroupedStats$rep2$five_prime,
       file.path(tableOutputDir, "CPD_mismatch_five_prime_position_stats_rep2.tsv"),sep = '\t')

fwrite(sixFourGroupedStats$rep1$three_prime,
       file.path(tableOutputDir, "6-4_mismatch_three_prime_position_stats_rep1.tsv"), sep = '\t')
fwrite(sixFourGroupedStats$rep2$three_prime,
       file.path(tableOutputDir, "6-4_mismatch_three_prime_position_stats_rep2.tsv"), sep = '\t')
fwrite(sixFourGroupedStats$rep1$five_prime,
       file.path(tableOutputDir, "6-4_mismatch_five_prime_position_stats_rep1.tsv"), sep = '\t')
fwrite(sixFourGroupedStats$rep2$five_prime,
       file.path(tableOutputDir, "6-4_mismatch_five_prime_position_stats_rep2.tsv"), sep = '\t')
```


```{r position_differences_across_read_lengths, fig.width = 9, fig.height = 5}
plotGroupedPositionStats(CPDGroupedStats$rep1$three_prime, CPDGroupedStats$rep1$five_prime, POS_DIFF,
                         title = "CPD C>T Mismatch Pos Diff Rep1",
                         xAxisBreaks = c(23:31))
plotGroupedPositionStats(CPDGroupedStats$rep2$three_prime, CPDGroupedStats$rep2$five_prime, POS_DIFF,
                         title = "CPD C>T Mismatch Pos Diff Rep2",
                         xAxisBreaks = c(23:31))

plotGroupedPositionStats(sixFourGroupedStats$rep1$three_prime, sixFourGroupedStats$rep1$five_prime, POS_DIFF,
                         title = "6-4 C>T Mismatch Pos Diff Rep1",
                         xAxisBreaks = c(23:31))
plotGroupedPositionStats(sixFourGroupedStats$rep2$three_prime, sixFourGroupedStats$rep2$five_prime, POS_DIFF,
                         title = "6-4 C>T Mismatch Pos Diff Rep2",
                         xAxisBreaks = c(23:31))
```


# ```{r plot_stats_across_read_lengths, fig.width = 9, fig.height = 5}
# plotGroupedPositionStats(CPDGroupedStats$rep1$three_prime, CPDGroupedStats$rep1$five_prime, IQR,
#                          title = "CPD C>T Mismatch Pos Diff Rep1",
#                          xAxisBreaks = c(23:31))
# plotGroupedPositionStats(CPDGroupedStats$rep2$three_prime, CPDGroupedStats$rep2$five_prime, IQR,
#                          title = "CPD C>T Mismatch Pos Diff Rep2",
#                          xAxisBreaks = c(23:31))
# 
# plotGroupedPositionStats(sixFourGroupedStats$rep1$three_prime, sixFourGroupedStats$rep1$five_prime, IQR,
#                          title = "6-4 C>T Mismatch Pos Diff Rep1",
#                          xAxisBreaks = c(23:31))
# plotGroupedPositionStats(sixFourGroupedStats$rep2$three_prime, sixFourGroupedStats$rep2$five_prime, IQR,
#                          title = "6-4 C>T Mismatch Pos Diff Rep2",
#                          xAxisBreaks = c(23:31))
# ```


```{r purine_vs_pyrimidine_frequencies_across_read_length, fig.width = 11, fig.height = 5}
plotNucFreqVsReadLengthBarPlot(CPDSingleCToTMismatchData$rep1, THREE_PRIME,
                               title = "CPD Rep1 3' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(CPDSingleCToTMismatchData$rep2, THREE_PRIME,
                               title = "CPD Rep2 3' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(CPDSingleCToTMismatchData$rep1, FIVE_PRIME,
                               title = "CPD Rep1 5' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(CPDSingleCToTMismatchData$rep2, FIVE_PRIME,
                               title = "CPD Rep2 5' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))

plotNucFreqVsReadLengthBarPlot(sixFourSingleCToTMismatchData$rep1, THREE_PRIME,
                               title = "6-4 Rep1 3' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(sixFourSingleCToTMismatchData$rep2, THREE_PRIME,
                               title = "6-4 Rep2 3' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(sixFourSingleCToTMismatchData$rep1, FIVE_PRIME,
                               title = "6-4 Rep1 5' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(sixFourSingleCToTMismatchData$rep2, FIVE_PRIME,
                               title = "6-4 Rep2 5' Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
```


```{r filter_by_mismatch_pos_and_read_length}
CPDFilteredSingleCToTMismatchData = lapply(CPDSingleCToTMismatchData, function(x) {
  lapply(x, filterMismatchesByPositionAndReadLength, HUMAN_THREE_PRIME_POS_CONSTRAINTS)
})
sixFourFilteredSingleCToTMismatchData = lapply(sixFourSingleCToTMismatchData, function(x) {
  lapply(x, filterMismatchesByPositionAndReadLength, HUMAN_THREE_PRIME_POS_CONSTRAINTS)
})
```



```{r purine_vs_pyrimidine_frequencies_first_quartile_mismatch_pos, fig.width = 11, fig.height = 5}

plotNucFreqVsReadLengthBarPlot(lapply(CPDFilteredSingleCToTMismatchData$rep1, getFirstPositionQuartile),
                               THREE_PRIME, title = "CPD Rep1 1st Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(lapply(CPDFilteredSingleCToTMismatchData$rep2, getFirstPositionQuartile),
                               THREE_PRIME, title = "CPD Rep2 1st Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))

plotNucFreqVsReadLengthBarPlot(lapply(sixFourFilteredSingleCToTMismatchData$rep1, getFirstPositionQuartile),
                               THREE_PRIME, title = "6-4 Rep1 1st Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(lapply(sixFourFilteredSingleCToTMismatchData$rep2, getFirstPositionQuartile),
                               THREE_PRIME, title = "6-4 Rep2 1st Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
```


```{r purine_vs_pyrimidine_frequencies_last_quartile_mismatch_pos, fig.width = 11, fig.height = 5}

plotNucFreqVsReadLengthBarPlot(lapply(CPDFilteredSingleCToTMismatchData$rep1, getLastPositionQuartile),
                               THREE_PRIME, title = "CPD Rep1 4th Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(lapply(CPDFilteredSingleCToTMismatchData$rep2, getLastPositionQuartile),
                               THREE_PRIME, title = "CPD Rep2 4th Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))

plotNucFreqVsReadLengthBarPlot(lapply(sixFourFilteredSingleCToTMismatchData$rep1, getLastPositionQuartile),
                               THREE_PRIME, title = "6-4 Rep1 4th Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
plotNucFreqVsReadLengthBarPlot(lapply(sixFourFilteredSingleCToTMismatchData$rep2, getLastPositionQuartile),
                               THREE_PRIME, title = "6-4 Rep2 4th Quartile Nuc Freq",
                               combineNucleotides = c("AG","CT"), combinedNames = c("Purine","Pyrimidine"))
```


# ```{r acquire_and_format_data_2}
# # Read in the GC content vs. read length data
# CPDReadLengthAndGCFilePaths = list(rep1 = list(), rep2 = list())
# CPDReadLengthAndGCFilePaths$rep1 = list(
#   hr1 = file.path(deaminationDataDirectory, "NHF1_CPD_1h_Rep1_read_length_vs_GC_content.tsv"),
#   hr8 = file.path(deaminationDataDirectory, "NHF1_CPD_8h_Rep1_read_length_vs_GC_content.tsv"),
#   hr24 = file.path(deaminationDataDirectory, "NHF1_CPD_24h_Rep1_read_length_vs_GC_content.tsv"),
#   hr48 = file.path(deaminationDataDirectory, "NHF1_CPD_48h_Rep1_read_length_vs_GC_content.tsv")
# )
# CPDReadLengthAndGCFilePaths$rep2 = list(
#   hr1 = file.path(deaminationDataDirectory, "NHF1_CPD_1h_Rep2_read_length_vs_GC_content.tsv"),
#   hr8 = file.path(deaminationDataDirectory, "NHF1_CPD_8h_Rep2_read_length_vs_GC_content.tsv"),
#   hr24 = file.path(deaminationDataDirectory, "NHF1_CPD_24h_Rep2_read_length_vs_GC_content.tsv"),
#   hr48 = file.path(deaminationDataDirectory, "NHF1_CPD_48h_Rep2_read_length_vs_GC_content.tsv")
# )
# 
# sixFourReadLengthAndGCFilePaths = list(rep1 = list(), rep2 = list())
# sixFourReadLengthAndGCFilePaths$rep1 = list(
#   min5 = file.path(deaminationDataDirectory, "NHF1_6-4-PP_5min_Rep1_read_length_vs_GC_content.tsv"),
#   min20 = file.path(deaminationDataDirectory, "NHF1_6-4-PP_20min_Rep1_read_length_vs_GC_content.tsv"),
#   hr1 = file.path(deaminationDataDirectory, "NHF1_6-4-PP_1h_Rep1_read_length_vs_GC_content.tsv")
# )
# sixFourReadLengthAndGCFilePaths$rep2 = list(
#   min5 = file.path(deaminationDataDirectory, "NHF1_6-4-PP_5min_Rep2_read_length_vs_GC_content.tsv"),
#   min20 = file.path(deaminationDataDirectory, "NHF1_6-4-PP_20min_Rep2_read_length_vs_GC_content.tsv"),
#   hr1 = file.path(deaminationDataDirectory, "NHF1_6-4-PP_1h_Rep2_read_length_vs_GC_content.tsv")
# )
# 
# CPDReadLengthAndGCData = lapply(CPDReadLengthAndGCFilePaths, function(x) lapply(x, fread))
# sixFourReadLengthAndGCData = lapply(sixFourReadLengthAndGCFilePaths, function(x) lapply(x, fread))
# ```


# ```{r CPD_read_length_vs_GC_content}
# for (i in 1:length(CPDReadLengthAndGCData)) {
#   for (j in 1:length(CPDReadLengthAndGCData[[i]])) {
#     plotGCContentVsReadLength(CPDReadLengthAndGCData[[i]][[j]],
#                               title = paste("CPD", names(CPDReadLengthAndGCData)[i], 
#                                             names(CPDReadLengthAndGCData[[i]])[j], "GC Content Vs Read Length"))
#   }
# }
# ```
# 
# 
# ```{r 6-4_read_length_vs_GC_content}
# for (i in 1:length(sixFourReadLengthAndGCData)) {
#   for (j in 1:length(sixFourReadLengthAndGCData[[i]])) {
#     plotGCContentVsReadLength(sixFourReadLengthAndGCData[[i]][[j]],
#                               title = paste("6-4", names(sixFourReadLengthAndGCData)[i], 
#                                             names(sixFourReadLengthAndGCData[[i]])[j], "GC Content Vs Read Length"))
#   }
# }
# ```


```{r acquire_and_format_data_3}
load(file.path(deaminationDataDirectory, "periodicity_results", "NHF1_deamination_periodicity_results.rda"))
dataSetNames = as.list(setNames(nm = mutperiodData$periodicityResults$Data_Set))
```


```{r CPD_rotational_periodicity_trinuc_normalized}
parseAndPlotPeriodicity(dataSetNames$NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_normalized,
                        title = "CPD 1hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(dataSetNames$NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_normalized,
                        title = "CPD 24hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(dataSetNames$NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_normalized,
                        title = "CPD 48hr Rep1 Rotational Trinuc Norm")
```


```{r CPD_rotational_periodicity_XR-seq_normalized}
parseAndPlotPeriodicity(
  dataSetNames$NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_normalized,
  title = "CPD 1hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_normalized,
  title = "CPD 24hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_normalized,
  title = "CPD 48hr Rep1 Rotational XR-seq Norm")
```


```{r CPD_translational_periodicity_trinuc_normalized}
parseAndPlotPeriodicity(
  dataSetNames$`NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_nuc-group_normalized`,
  title = "CPD 1hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(
  dataSetNames$`NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_nuc-group_normalized`,
  title = "CPD 24hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(
  dataSetNames$`NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_nuc-group_normalized`,
  title = "CPD 48hr Rep1 Rotational Trinuc Norm")
```



```{r CPD_translational_periodicity_XR-seq_normalized}
parseAndPlotPeriodicity(
  dataSetNames$`NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_nuc-group_normalized`,
  title = "CPD 1hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$`NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_nuc-group_normalized`,
  title = "CPD 24hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$`NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_nuc-group_normalized`,
  title = "CPD 48hr Rep1 Rotational XR-seq Norm")
```


```{r acquire_and_format_data_4}
load(file.path(deaminationDataDirectory, "periodicity_results", "NHF1_TCG_deamination_periodicity_results.rda"))
dataSetNames = as.list(setNames(nm = mutperiodData$periodicityResults$Data_Set))

```


```{r TCG_CPD_rotational_periodicity_trinuc_normalized}
parseAndPlotPeriodicity(
  dataSetNames$TCG_NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_normalized,
  title = "TCG CPD 1hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(
  dataSetNames$TCG_NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_normalized,
  title = "TCG CPD 24hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(
  dataSetNames$TCG_NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_normalized,
  title = "TCG CPD 48hr Rep1 Rotational Trinuc Norm")
```


```{r TCG_CPD_rotational_periodicity_XR-seq_normalized}
parseAndPlotPeriodicity(
  dataSetNames$TCG_NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_normalized,
  title = "TCG CPD 1hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$TCG_NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_normalized,
  title = "TCG CPD 24hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$TCG_NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_normalized,
  title = "TCG CPD 48hr Rep1 Rotational XR-seq Norm")
```


```{r TCG_CPD_translational_periodicity_trinuc_normalized}
parseAndPlotPeriodicity(
  dataSetNames$`TCG_NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_nuc-group_normalized`,
  title = "TCG CPD 1hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(
  dataSetNames$`TCG_NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_nuc-group_normalized`,
  title = "TCG CPD 24hr Rep1 Rotational Trinuc Norm")
parseAndPlotPeriodicity(
  dataSetNames$`TCG_NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_trinuc_nuc-group_normalized`,
  title = "TCG CPD 48hr Rep1 Rotational Trinuc Norm")
```



```{r TCG_CPD_translational_periodicity_XR-seq_normalized}
parseAndPlotPeriodicity(
  dataSetNames$`TCG_NHF1_CPD_1h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_nuc-group_normalized`,
  title = "TCG_CPD 1hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$`TCG_NHF1_CPD_24h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_nuc-group_normalized`,
  title = "TCG CPD 24hr Rep1 Rotational XR-seq Norm")
parseAndPlotPeriodicity(
  dataSetNames$`TCG_NHF1_CPD_48h_Rep1_individual_mismatches_MNase_map_intergenic_custom_context_nuc-group_normalized`,
  title = "TCG CPD 48hr Rep1 Rotational XR-seq Norm")
```
