---
title: "Human Deamination Read Length Analysis"
author: "Ben Morledge-Hampton"
date: "5/2/2022"
output: html_document
params:
  bioinformaticsDirectory:
    label: "Bioinformatics Projects Directory"
    value: C:/Bioinformatics_Projects
    input: select
    choices: [C:/Bioinformatics_Projects, F:/Bioinformatics_Projects]
---


```{r setup_chunk}
library(knitr)
opts_chunk$set(dev = "png", dpi = 600, fig.width = 7, fig.height = 5, 
               fig.path = paste0(file.path("deamination_analysis_output","human_analysis_figures"),
                                 .Platform$file.sep))
if (interactive()) {
  bioinformaticsDirectory = choose.dir(caption = "Select Bioinformatics_Projects Directory")
} else {
  bioinformaticsDirectory = params$bioinformaticsDirectory
}

deaminationDataDirectory = file.path(bioinformaticsDirectory, "deamination_determination", "data", "human")
```


```{r source_functions}
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R', "DeaminationAnalysis.R"))
```


```{r get_file_paths}
expansionOffset = 10

# Get file paths for the expanded mismatch by read data
expandedCPDMismatchDataFilePaths = list(rep1 = list(), rep2 = list())
expandedCPDMismatchDataFilePaths$rep1 = list(
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read",
                  paste0("NHF1_CPD_1h_Rep1_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr8 = file.path(deaminationDataDirectory, "mismatches_by_read",
                  paste0("NHF1_CPD_8h_Rep1_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr24 = file.path(deaminationDataDirectory, "mismatches_by_read",
                   paste0("NHF1_CPD_24h_Rep1_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr48 = file.path(deaminationDataDirectory, "mismatches_by_read",
                   paste0("NHF1_CPD_48h_Rep1_mismatches_by_read_",expansionOffset,"bp_expanded.bed"))
)
expandedCPDMismatchDataFilePaths$rep2 = list(
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read",
                  paste0("NHF1_CPD_1h_Rep2_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr8 = file.path(deaminationDataDirectory, "mismatches_by_read",
                  paste0("NHF1_CPD_8h_Rep2_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr24 = file.path(deaminationDataDirectory, "mismatches_by_read",
                   paste0("NHF1_CPD_24h_Rep2_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr48 = file.path(deaminationDataDirectory, "mismatches_by_read",
                   paste0("NHF1_CPD_48h_Rep2_mismatches_by_read_",expansionOffset,"bp_expanded.bed"))
)

expandedSixFourMismatchDataFilePaths = list(rep1 = list(), rep2 = list())
expandedSixFourMismatchDataFilePaths$rep1 = list(
  min5 = file.path(deaminationDataDirectory, "mismatches_by_read",
                   paste0("NHF1_6-4-PP_5min_Rep1_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  min20 = file.path(deaminationDataDirectory, "mismatches_by_read",
                    paste0("NHF1_6-4-PP_20min_Rep1_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read",
                  paste0("NHF1_6-4-PP_1h_Rep1_mismatches_by_read_",expansionOffset,"bp_expanded.bed"))
)
expandedSixFourMismatchDataFilePaths$rep2 = list(
  min5 = file.path(deaminationDataDirectory, "mismatches_by_read",
                   paste0("NHF1_6-4-PP_5min_Rep2_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  min20 = file.path(deaminationDataDirectory, "mismatches_by_read",
                    paste0("NHF1_6-4-PP_20min_Rep2_mismatches_by_read_",expansionOffset,"bp_expanded.bed")),
  hr1 = file.path(deaminationDataDirectory, "mismatches_by_read",
                  paste0("NHF1_6-4-PP_1h_Rep2_mismatches_by_read_",expansionOffset,"bp_expanded.bed"))
)
```


```{r acquire_and_format_data}
expandedCPDSingleCToTMismatchData = lapply(expandedCPDMismatchDataFilePaths, function(x) lapply(x, fread))
expandedSixFourSingleCToTMismatchData = lapply(expandedSixFourMismatchDataFilePaths, function(x) lapply(x, fread))

# Create a subset of single mismatch C>T reads.
expandedCPDSingleCToTMismatchData = lapply(expandedCPDSingleCToTMismatchData, function(x) lapply(
  x, function(y) filterResults(y[V5 == "C>T"], minReadLength = 23, maxReadLength = 31)
))
expandedSixFourSingleCToTMismatchData = lapply(expandedSixFourSingleCToTMismatchData, function(x) lapply(
  x, function(y) filterResults(y[V5 == "C>T"], minReadLength = 23, maxReadLength = 31)
))

# Simplify the data
expandedCPDSingleCToTMismatchData = lapply(expandedCPDSingleCToTMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y, TRUE, expansionOffset = expansionOffset)
))
expandedSixFourSingleCToTMismatchData = lapply(expandedSixFourSingleCToTMismatchData, function(x) lapply(
  x, function(y) simplifyTable(y, TRUE, expansionOffset = expansionOffset)
))

# Further filter data by read position based on length
expandedCPDSingleCToTMismatchData = lapply(expandedCPDSingleCToTMismatchData, function(x) {
  lapply(x, filterMismatchesByPositionAndReadLength, HUMAN_THREE_PRIME_POS_CONSTRAINTS, expansionOffset)
})

expandedSixFourSingleCToTMismatchData = lapply(expandedSixFourSingleCToTMismatchData, function(x) {
  lapply(x, filterMismatchesByPositionAndReadLength, HUMAN_THREE_PRIME_POS_CONSTRAINTS, expansionOffset)
})
```


```{r orient_data_to_mismatch_positions}
threePrimeMismatchAnchoredCPDData = lapply(expandedCPDSingleCToTMismatchData, function(x) {
  lapply(x, orientDataToMismatch, THREE_PRIME, expansionOffset)
})
fivePrimeMismatchAnchoredCPDData = lapply(expandedCPDSingleCToTMismatchData, function(x) {
  lapply(x, orientDataToMismatch, FIVE_PRIME, expansionOffset)
})

threePrimeMismatchAnchoredSixFourData = lapply(expandedSixFourSingleCToTMismatchData, function(x) {
  lapply(x, orientDataToMismatch, THREE_PRIME, expansionOffset)
})
fivePrimeMismatchAnchoredSixFourData = lapply(expandedSixFourSingleCToTMismatchData, function(x) {
  lapply(x, orientDataToMismatch, FIVE_PRIME, expansionOffset)
})
```


```{r generate_nucleotide_frequency_data}
threePrimeMismatchCPDNucFreq = lapply(threePrimeMismatchAnchoredCPDData, tabulateNucFreqByPosTimepointAndLength)
fivePrimeMismatchCPDNucFreq = lapply(fivePrimeMismatchAnchoredCPDData,
                                     tabulateNucFreqByPosTimepointAndLength, posType = FIVE_PRIME)

threePrimeMismatchSixFourNucFreq = lapply(threePrimeMismatchAnchoredSixFourData,
                                          tabulateNucFreqByPosTimepointAndLength)
fivePrimeMismatchSixFourNucFreq = lapply(fivePrimeMismatchAnchoredSixFourData,
                                         tabulateNucFreqByPosTimepointAndLength, posType = FIVE_PRIME)
```


```{r CPD_mismatch__anchored_nucleotide_frequencies, fig.width = 11, fig.height = 7}
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchCPDNucFreq$rep1, 
                               title = "CPD 3' Anchored Nucleotide Frequencies Rep1",
                               secondaryYAxisLabel = "5' Cut Site Distance",
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchCPDNucFreq$rep2, 
                               title = "CPD 3' Anchored Nucleotide Frequencies Rep2",
                               secondaryYAxisLabel = "5' Cut Site Distance",
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)

plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchCPDNucFreq$rep1, 
                               title = "CPD 5' Anchored Nucleotide Frequencies Rep1",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchCPDNucFreq$rep2, 
                               title = "CPD 5' Anchored Nucleotide Frequencies Rep2",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
```



```{r CPD_mismatch_anchored_purine_frequencies, fig.width = 11, fig.height = 7}
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchCPDNucFreq$rep1, 
                               title = "CPD 3' Anchored Purine Frequencies Rep1",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "5' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchCPDNucFreq$rep2, 
                               title = "CPD 3' Anchored Purine Frequencies Rep2",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "5' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)

plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchCPDNucFreq$rep1, 
                               title = "CPD 5' Anchored Purine Frequencies Rep1",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchCPDNucFreq$rep2, 
                               title = "CPD 5' Anchored Purine Frequencies Rep2",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
```


```{r CPD_mismatch__anchored_nucleotide_frequencies, fig.width = 11, fig.height = 7}
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchSixFourNucFreq$rep1, 
                               title = "6-4 3' Anchored Nucleotide Frequencies Rep1",
                               secondaryYAxisLabel = "5' Cut Site Distance",
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchSixFourNucFreq$rep2, 
                               title = "6-4 3' Anchored Nucleotide Frequencies Rep2",
                               secondaryYAxisLabel = "5' Cut Site Distance",
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)

plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchSixFourNucFreq$rep1, 
                               title = "6-4 5' Anchored Nucleotide Frequencies Rep1",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchSixFourNucFreq$rep2, 
                               title = "6-4 5' Anchored Nucleotide Frequencies Rep2",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
```


```{r CPD_mismatch_anchored_purine_frequencies, fig.width = 11, fig.height = 7}
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchSixFourNucFreq$rep1, 
                               title = "6-4 3' Anchored Purine Frequencies Rep1",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "5' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(threePrimeMismatchSixFourNucFreq$rep2, 
                               title = "6-4 3' Anchored Purine Frequencies Rep2",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "5' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = THREE_PRIME, showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset)

plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchSixFourNucFreq$rep1, 
                               title = "6-4 5' Anchored Purine Frequencies Rep1",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
plotNucFreqVsReadLengthBarPlot(fivePrimeMismatchSixFourNucFreq$rep2, 
                               title = "6-4 5' Anchored Purine Frequencies Rep2",
                               yAxisLabel = "Purine Frequency", secondaryYAxisLabel = "3' Cut Site Distance",
                               combineNucleotides = c("AG"), combinedNames = c("Purine"),
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset)
```


```{r compute_sequence_enrichment}
threePrimeMMAnchoredCPDEnrichment = mapply(function(mismatchDataRep, nucFreqRep) {
  setNames(lapply(seq_along(mismatchDataRep), function(i) {
    getSequenceEnrichmentByReadLengthAndPos(mismatchDataRep[[i]],
                                            nucFreqRep[Timepoint == names(mismatchDataRep)[i]],
                                            c("CG","CA","TG","TA"))
  }), names(mismatchDataRep))
}, threePrimeMismatchAnchoredCPDData, threePrimeMismatchCPDNucFreq, SIMPLIFY = FALSE)

fivePrimeMMAnchoredCPDEnrichment = mapply(function(mismatchDataRep, nucFreqRep) {
  setNames(lapply(seq_along(mismatchDataRep), function(i) {
    getSequenceEnrichmentByReadLengthAndPos(mismatchDataRep[[i]],
                                            nucFreqRep[Timepoint == names(mismatchDataRep)[i]],
                                            c("TGG"), posType = FIVE_PRIME)
  }), names(mismatchDataRep))
}, fivePrimeMismatchAnchoredCPDData, fivePrimeMismatchCPDNucFreq, SIMPLIFY = FALSE)


threePrimeMMAnchoredSixFourEnrichment = mapply(function(mismatchDataRep, nucFreqRep) {
  setNames(lapply(seq_along(mismatchDataRep), function(i) {
    getSequenceEnrichmentByReadLengthAndPos(mismatchDataRep[[i]],
                                            nucFreqRep[Timepoint == names(mismatchDataRep)[i]],
                                            c("CG","CA","TG","TA"))
  }), names(mismatchDataRep))
}, threePrimeMismatchAnchoredSixFourData, threePrimeMismatchSixFourNucFreq, SIMPLIFY = FALSE)

fivePrimeMMAnchoredSixFourEnrichment = mapply(function(mismatchDataRep, nucFreqRep) {
  setNames(lapply(seq_along(mismatchDataRep), function(i) {
    getSequenceEnrichmentByReadLengthAndPos(mismatchDataRep[[i]],
                                            nucFreqRep[Timepoint == names(mismatchDataRep)[i]],
                                            c("TGG"), posType = FIVE_PRIME)
  }), names(mismatchDataRep))
}, fivePrimeMismatchAnchoredSixFourData, fivePrimeMismatchSixFourNucFreq, SIMPLIFY = FALSE)
```


```{r plot_CPD_sequence_enrichment, fig.width = 11, fig.height = 7}
plotSequenceEnrichment(threePrimeMMAnchoredCPDEnrichment$rep1,
                       title = "CPD 3' Anchored YR Enrichment Rep1", secondaryYAxisLabel = "5' Cut Site Distance",
                       showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset,
                       querySequences = c("CG","CA","TG","TA"))
plotSequenceEnrichment(threePrimeMMAnchoredCPDEnrichment$rep2,
                       title = "CPD 3' Anchored YR Enrichment Rep2", secondaryYAxisLabel = "5' Cut Site Distance",
                       showFivePrimeCutSite = TRUE, expansionOffset = expansionOffset,
                       querySequences = c("CG","CA","TG","TA"))

plotSequenceEnrichment(fivePrimeMMAnchoredCPDEnrichment$rep1,
                       title = "CPD 5' Anchored TGG Enrichment Rep1", secondaryYAxisLabel = "3' Cut Site Distance",
                       showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset, posType = FIVE_PRIME,
                       querySequences = c("TGG"))
plotSequenceEnrichment(fivePrimeMMAnchoredCPDEnrichment$rep2,
                       title = "CPD 5' Anchored TGG Enrichment Rep2", secondaryYAxisLabel = "3' Cut Site Distance",
                       showThreePrimeCutSite = TRUE, expansionOffset = expansionOffset, posType = FIVE_PRIME,
                       querySequences = c("TGG"))
```


```{r enrichment_test_1}
for (i in 1:5) {
  testTable = data.table(Read_Length = 9,
                         Read_Sequence = replicate(1000, paste0(sample(c('A','C','G','T'), 9, TRUE), collapse = '')))
  testTableNucFreq = tabulateNucFreqByPosTimepointAndLength(testTable)
  testEnrichment = getSequenceEnrichmentByReadLengthAndPos(testTable, testTableNucFreq, c("CG","CA","TG","TA"))
  plotSequenceEnrichment(testEnrichment, title = "CPD 3' Anchored YR Enrichment Rep1")
}
```


```{r enrichment_test_2}
for (i in 1:5) {
  testTable = data.table(Read_Length = 9,
                         Read_Sequence = replicate(1000,
                                                   paste0(sample(c(rep('A',6),rep('C',4),rep('G',4),rep('T',6)),
                                                                 9, TRUE), collapse = '')))
  testTableNucFreq = tabulateNucFreqByPosTimepointAndLength(testTable)
  testEnrichment = getSequenceEnrichmentByReadLengthAndPos(testTable, testTableNucFreq, c("CG","CA","TG","TA"))
  plotSequenceEnrichment(testEnrichment, title = "CPD 3' Anchored YR Enrichment Rep1")
}
```
