---
title: "Paper Figure Generation"
author: "Ben Morledge-Hampton"
date: "6/21/2022"
output: html_document
params:
  bioinformaticsDirectory:
    label: "Bioinformatics Projects Directory"
    value: C:/Bioinformatics_Projects
    input: select
    choices: [C:/Bioinformatics_Projects, F:/Bioinformatics_Projects]
---

```{r setup_chunk}
library(knitr)

NHF1CPDTimepoints = list(`1h` = NULL, `8h` = NULL, `24h` = NULL, `48h` = NULL)
NHF1SixFourTimepoints = list(`5min` = NULL, `20min` = NULL, `1h` = NULL)
arabidopsisCPDTimepoints = list(`ZT2` = NULL, `ZT14` = NULL)
yeastCPDTimepoints = list(`5min` = NULL, `20min` = NULL, `1h` = NULL)
GM12878CisplatinTimepoints = list(`90min` = NULL)
CSBCPDTimepoints = list(`1h` = NULL)
XPCCPDTimepoints = list(`1h` = NULL)
EColiCPDTimepoints = list(`5min` = NULL)

if (interactive()) {
  bioinformaticsDirectory = choose.dir(caption = "Select Bioinformatics_Projects Directory")
} else {
  bioinformaticsDirectory = params$bioinformaticsDirectory
}

opts_chunk$set(dev = "png", dpi = 600, fig.width = 7, fig.height = 5, 
               fig.path = paste0(file.path("deamination_analysis_output","paper_figures"),
                                 .Platform$file.sep))

deaminationDataDirectory = file.path(bioinformaticsDirectory, "deamination_determination", "data")

NHF1DataDirectory = file.path(deaminationDataDirectory, "NHF1")
NHF1MismatchesByReadDataDirectory = file.path(NHF1DataDirectory, "mismatches_by_read")
NHF1SequenceAndPositionDataDirectory = file.path(NHF1DataDirectory, "sequence_and_position_analysis")
NHF1ExcisionSiteDataDirectory = file.path(NHF1DataDirectory, "excision_site_analysis")
NHF1GeneralReadSequenceDataDirectory = file.path(NHF1DataDirectory, "general_read_sequences")

arabidopsisDataDirectory = file.path(deaminationDataDirectory, "Arabidopsis")
arabidopsisSequenceAndPositionDataDirectory = file.path(arabidopsisDataDirectory, "sequence_and_position_analysis")
arabidopsisExcisionSiteDataDirectory = file.path(arabidopsisDataDirectory, "excision_site_analysis")
arabidopsisGeneralReadSequenceDataDirectory = file.path(arabidopsisDataDirectory, "general_read_sequences")

yeastDataDirectory = file.path(deaminationDataDirectory, "yeast")
yeastSequenceAndPositionDataDirectory = file.path(yeastDataDirectory, "sequence_and_position_analysis")
yeastExcisionSiteDataDirectory = file.path(yeastDataDirectory, "excision_site_analysis")
yeastGeneralReadSequenceDataDirectory = file.path(yeastDataDirectory, "general_read_sequences")

GM12878DataDirectory = file.path(deaminationDataDirectory, "GM12878")
GM12878SequenceAndPositionDataDirectory = file.path(GM12878DataDirectory, "sequence_and_position_analysis")
GM12878GeneralReadSequenceDataDirectory = file.path(GM12878DataDirectory, "general_read_sequences")

CSBDataDirectory = file.path(deaminationDataDirectory, "CS-B")
CSBSequenceAndPositionDataDirectory = file.path(CSBDataDirectory, "sequence_and_position_analysis")

XPCDataDirectory = file.path(deaminationDataDirectory, "XP-C")
XPCSequenceAndPositionDataDirectory = file.path(XPCDataDirectory, "sequence_and_position_analysis")

EColiDataDirectory = file.path(deaminationDataDirectory, "E_coli_WT")
EColiExcisionSiteDataDirectory = file.path(EColiDataDirectory, "excision_site_analysis")
EColiGeneralReadSequenceDataDirectory = file.path(EColiDataDirectory, "general_read_sequences")
```


```{r source_functions}
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "UsefulFileSystemFunctions.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationParseAndFilter.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationSequenceAndPositionAnalysis.R"))
source(file.path(bioinformaticsDirectory, "deamination_determination", 'R',
                 "DeaminationAnchoredExcisionSiteAnalysis.R"))
```


```{r read_figure_1_data}
NHF1CPDMismatchDataAllReps = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDMismatchDataAllReps)) {
  NHF1CPDMismatchDataAllReps[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    timepoint = names(NHF1CPDMismatchDataAllReps)[i]
  ))
}
```


```{r figure_1B}
plotMismatchTypeFrequencies(NHF1CPDMismatchDataAllReps$`1h`, ylim = c(0,0.8),
                            title = "NHF1 CPD 1h All Reps Mismatch Type Frequencies")
```
```{r figure_1C}
print("NHF1 CPD 1h all reps C>T mismatch dipyrimidine context frequency:")
plotTrinucleotideContext(NHF1CPDMismatchDataAllReps$`1h`, includedTypes = c("C>T"),
                         title = "NHF1 CPD 1h All Reps C>T Mismatch Trinuc Context",
                         printDipyFreq = TRUE)
```
```{r figure_1D}
plotMismatchPositionFrequencies(NHF1CPDMismatchDataAllReps$`1h`, includedTypes = c("C>T"),
                                title = "NHF1 CPD 1h All Reps C>T Mismatch Pos Frequencies",
                                xAxisBreaks = c(-1,-10,-20,-30))
```


```{r read_figure_2_data}
NHF1CPDMMFreqZScoresAllReps = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDMMFreqZScoresAllReps)) {
  NHF1CPDMMFreqZScoresAllReps[[i]] = fread(generateFilePath(
    directory = NHF1SequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    timepoint = names(NHF1CPDMMFreqZScoresAllReps)[i], includedMismatches = c("C>T")
  ))
}

NHF1FivePrimeMismatchCPDNucFreqAllReps = fread(generateFilePath(
  directory = NHF1ExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
  strandPolarity = "five_prime", anchored = TRUE
))

NHF1CPDTGGSeqFreqAllReps = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDTGGSeqFreqAllReps)) {
  NHF1CPDTGGSeqFreqAllReps[[i]] = fread(generateFilePath(
    directory = NHF1GeneralReadSequenceDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    timepoint = names(NHF1CPDTGGSeqFreqAllReps)[i], sequence = "TGG"
  ))
}
```


```{r figure_2A, fig.width = 5, fig.height = 6}
### Just the 1h timepoint ###
plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllReps["1h"], includedTypes = c("C>T"),
                                         title = "NHF1 CPD 1h All Reps C>T Mismatch Pos Freq",
                                         yAxisLabel = "Relative C>T Frequency",
                                         zScoreTables = NHF1CPDMMFreqZScoresAllReps,
                                         xAxisBreaks = c(-1,-10,-20,-30))

### All timepoints together ###
# plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllReps, includedTypes = c("C>T"),
#                                          title = "NHF1 CPD All Reps C>T Mismatch Pos Freq",
#                                          yAxisLabel = "Relative C>T Frequency",
#                                          zScoreTables = NHF1CPDMMFreqZScoresAllReps)
```
```{r figure_2B, fig.width = 6.5, fig.height = 6}
### Just the 1h timepoint ###
plotNucFreqVsReadLengthBarPlot(NHF1FivePrimeMismatchCPDNucFreqAllReps[Timepoint == "1h"], 
                               title = "NHF1 CPD 1h All Reps 5' Anchored Nuc Freq",
                               secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE,
                               expansionOffset = 10, minReadLength = 1, maxReadLength = 8,
                               xAxisLabel = "Mismatch Relative Position", positionOffset = -1)

### All timepoints together ###
# plotNucFreqVsReadLengthBarPlot(NHF1FivePrimeMismatchCPDNucFreqAllReps, 
#                                title = "NHF1 CPD All Reps 5' Anchored Nuc Freq",
#                                secondaryYAxisLabel = "3' Cut Site Distance",
#                                posType = FIVE_PRIME, showThreePrimeCutSite = TRUE,
#                                expansionOffset = 10, minReadLength = 1, maxReadLength = 8,
#                                xAxisLabel = "Mismatch Relative Position")
```
```{r figure_2C, fig.width = 5, fig.height = 6}
### Just the 1h timepoint ###
plotSequenceFrequencies(NHF1CPDTGGSeqFreqAllReps$`1h`, title = "NHF1 CPD 1h All Reps General Read TGG Freq",
                        showThreePrimeCutSite = TRUE, showFivePrimeCutSite = TRUE, querySequences = c("TGG"),
                        expansionOffset = 10, minReadLength = 20, maxReadLength = 27, positionOffset = 10,
                        yAxisLabel = "TGG Trinucleotide Frequency", defaultColor = "Black")


### All timepoints together ###
# plotSequenceFrequencies(NHF1CPDTGGSeqFreqAllReps, title = "NHF1 CPD All Reps General Read TGG Freq",
#                         showThreePrimeCutSite = TRUE, showFivePrimeCutSite = TRUE, querySequences = c("TGG"),
#                         expansionOffset = 10, minReadLength = 21, maxReadLength = 27,
#                         yAxisLabel = "TGG Trinucleotide Frequency", defaultColor = "Black")
```


```{r read_figure_3_data}
NHF1CPDMismatchDataAllRepsTGGFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDMismatchDataAllRepsTGGFiltered)) {
  NHF1CPDMismatchDataAllRepsTGGFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "NHF1", lesion = "CPD", timepoint = names(NHF1CPDMismatchDataAllRepsTGGFiltered)[i],
    repitition = "all_reps", filtering = "TGG_filtered"
  ))
}

arabidopsisCPDMismatchDataAllRepsFiltered = arabidopsisCPDTimepoints
for (i in seq_along(arabidopsisCPDMismatchDataAllRepsFiltered)) {
  arabidopsisCPDMismatchDataAllRepsFiltered[[i]] = fread(generateFilePath(
    arabidopsisSequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "Arabidopsis", lesion = "CPD", timepoint = names(arabidopsisCPDMismatchDataAllRepsFiltered)[i],
    repitition = "all_reps", filtering = "TGG_filtered"
  ))
}

yeastCPDMismatchDataAllRepsFiltered = yeastCPDTimepoints
for (i in seq_along(yeastCPDMismatchDataAllRepsFiltered)) {
  yeastCPDMismatchDataAllRepsFiltered[[i]] = fread(generateFilePath(
    yeastSequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "yeast", lesion = "CPD", timepoint = names(yeastCPDMismatchDataAllRepsFiltered)[i],
    repitition = "all_reps", filtering = "TGG_filtered"
  ))
}

NHF1CPDMMFreqZScoresAllRepsTGGFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDMMFreqZScoresAllRepsTGGFiltered)) {
  NHF1CPDMMFreqZScoresAllRepsTGGFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    timepoint = names(NHF1CPDMMFreqZScoresAllRepsTGGFiltered)[i], includedMismatches = c("C>T")
  ))
}

arabidopsisCPDMMFreqZScoresAllRepsFiltered = arabidopsisCPDTimepoints
for (i in seq_along(arabidopsisCPDMMFreqZScoresAllRepsFiltered)) {
  arabidopsisCPDMMFreqZScoresAllRepsFiltered[[i]] = fread(generateFilePath(
    arabidopsisSequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
    cellType = "Arabidopsis", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    timepoint = names(arabidopsisCPDMMFreqZScoresAllRepsFiltered)[i], includedMismatches = c("C>T")
  ))
}

yeastCPDMMFreqZScoresAllRepsFiltered = yeastCPDTimepoints
for (i in seq_along(yeastCPDMMFreqZScoresAllRepsFiltered)) {
  yeastCPDMMFreqZScoresAllRepsFiltered[[i]] = fread(generateFilePath(
    yeastSequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
    cellType = "yeast", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    timepoint = names(yeastCPDMMFreqZScoresAllRepsFiltered)[i], includedMismatches = c("C>T")
  ))
}

NHF1CPDGroupedStatsAllRepsFiltered = list(three_prime = NULL, five_prime = NULL)
for (i in seq_along(NHF1CPDGroupedStatsAllRepsFiltered)) {
  NHF1CPDGroupedStatsAllRepsFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    strandPolarity = names(NHF1CPDGroupedStatsAllRepsFiltered)[i]
  ))
}

arabidopsisCPDGroupedStatsAllRepsFiltered = list(three_prime = NULL, five_prime = NULL)
for (i in seq_along(arabidopsisCPDGroupedStatsAllRepsFiltered)) {
  arabidopsisCPDGroupedStatsAllRepsFiltered[[i]] = fread(generateFilePath(
    arabidopsisSequenceAndPositionDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
    cellType = "Arabidopsis", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    strandPolarity = names(arabidopsisCPDGroupedStatsAllRepsFiltered)[i]
  ))
}

yeastCPDGroupedStatsAllRepsFiltered = list(three_prime = NULL, five_prime = NULL)
for (i in seq_along(yeastCPDGroupedStatsAllRepsFiltered)) {
  yeastCPDGroupedStatsAllRepsFiltered[[i]] = fread(generateFilePath(
    yeastSequenceAndPositionDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
    cellType = "yeast", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    strandPolarity = names(yeastCPDGroupedStatsAllRepsFiltered)[i]
  ))
}
```


```{r figure_3A, fig.width = 5, fig.height = 5}
### Just the 1h timepoint ###
plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGGFiltered["1h"], includedTypes = c("C>T"),
                                         title = "NHF1 CPD 1h All Reps TGG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGGFiltered,
                                         yAxisLabel = "Relative C>T Frequency", xAxisBreaks = c(-1,-10,-20,-30))
plotPositionAcrossTimepointAndReadLength(arabidopsisCPDMismatchDataAllRepsFiltered["ZT2"], includedTypes = c("C>T"),
                                         title = "Arabidopsis CPD ZT2 All Reps TGG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = arabidopsisCPDMMFreqZScoresAllRepsFiltered,
                                         yAxisLabel = "Relative C>T Frequency", xAxisBreaks = c(-1,-10,-20,-30))
plotPositionAcrossTimepointAndReadLength(yeastCPDMismatchDataAllRepsFiltered["20min"], includedTypes = c("C>T"),
                                         title = "Yeast CPD 20min All Reps TGG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = yeastCPDMMFreqZScoresAllRepsFiltered,
                                         yAxisLabel = "Relative C>T Frequency", xAxisBreaks = c(-1,-10,-20,-30))


### All timepoints together ###
# plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGGFiltered, includedTypes = c("C>T"),
#                                          title = "NHF1 CPD All Reps TGG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGGFiltered,
#                                          yAxisLabel = "Relative C>T Frequency")
# plotPositionAcrossTimepointAndReadLength(arabidopsisCPDMismatchDataAllRepsFiltered, includedTypes = c("C>T"),
#                                          title = "Arabidopsis CPD All Reps TGG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = arabidopsisCPDMMFreqZScoresAllRepsFiltered,
#                                          yAxisLabel = "Relative C>T Frequency")
# plotPositionAcrossTimepointAndReadLength(yeastCPDMismatchDataAllRepsFiltered, includedTypes = c("C>T"),
#                                          title = "Yeast CPD All Reps TGG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = yeastCPDMMFreqZScoresAllRepsFiltered,
#                                          yAxisLabel = "Relative C>T Frequency")
```
```{r figure_3B, fig.width = 5, fig.height = 5}
### Just the 1h timepoint ###
plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGGFiltered["1h"], includedTypes = c("C>T"),
                                         title = "NHF1 CPD 1h All Reps TGG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGGFiltered,
                                         posType = FIVE_PRIME, yAxisLabel = "Relative C>T Frequency",
                                         xAxisBreaks = c(1,10,20,30))
plotPositionAcrossTimepointAndReadLength(arabidopsisCPDMismatchDataAllRepsFiltered["ZT2"], includedTypes = c("C>T"),
                                         title = "Arabidopsis CPD ZT2 All Reps TGG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = arabidopsisCPDMMFreqZScoresAllRepsFiltered,
                                         posType = FIVE_PRIME, yAxisLabel = "Relative C>T Frequency",
                                         xAxisBreaks = c(1,10,20,30))
plotPositionAcrossTimepointAndReadLength(yeastCPDMismatchDataAllRepsFiltered["20min"], includedTypes = c("C>T"),
                                         title = "Yeast CPD 20min All Reps TGG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = yeastCPDMMFreqZScoresAllRepsFiltered,
                                         posType = FIVE_PRIME, yAxisLabel = "Relative C>T Frequency",
                                         xAxisBreaks = c(1,10,20,30))


### All timepoints together ###
# plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGGFiltered, includedTypes = c("C>T"),
#                                          title = "NHF1 CPD All Reps TGG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGGFiltered,
#                                          posType = FIVE_PRIME, yAxisLabel = "Relative C>T Frequency")
# plotPositionAcrossTimepointAndReadLength(arabidopsisCPDMismatchDataAllRepsFiltered, includedTypes = c("C>T"),
#                                          title = "Arabidopsis CPD All Reps TGG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = arabidopsisCPDMMFreqZScoresAllRepsFiltered,
#                                          posType = FIVE_PRIME, yAxisLabel = "Relative C>T Frequency")
# plotPositionAcrossTimepointAndReadLength(yeastCPDMismatchDataAllRepsFiltered, includedTypes = c("C>T"),
#                                          title = "Yeast CPD All Reps TGG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = yeastCPDMMFreqZScoresAllRepsFiltered,
#                                          posType = FIVE_PRIME, yAxisLabel = "Relative C>T Frequency")
```
```{r figure_3C, fig.width = 7.5, fig.height = 5}
### Just the 1h timepoint ###
plotGroupedPositionStats(NHF1CPDGroupedStatsAllRepsFiltered$three_prime[Timepoint == "1h"],
                         NHF1CPDGroupedStatsAllRepsFiltered$five_prime[Timepoint == "1h"],
                         POS_DIFF, title = "NHF1 CPD 1h All Reps TGG-Filtered C>T Mismatch Pos Diff",
                         xAxisBreaks = c(22:30))

plotGroupedPositionStats(arabidopsisCPDGroupedStatsAllRepsFiltered$three_prime[Timepoint == "ZT2"],
                         arabidopsisCPDGroupedStatsAllRepsFiltered$five_prime[Timepoint == "ZT2"],
                         POS_DIFF, title = "Arabidopsis CPD 1h All Reps TGG-Filtered C>T Mismatch Pos Diff",
                         xAxisBreaks = c(24:30))

plotGroupedPositionStats(yeastCPDGroupedStatsAllRepsFiltered$three_prime[Timepoint == "20min"],
                         yeastCPDGroupedStatsAllRepsFiltered$five_prime[Timepoint == "20min"],
                         POS_DIFF, title = "Yeast CPD 1h All Reps TGG-Filtered C>T Mismatch Pos Diff",
                         xAxisBreaks = c(22:25))


### All timepoints together ###
# plotGroupedPositionStats(NHF1CPDGroupedStatsAllRepsFiltered$three_prime,
#                          NHF1CPDGroupedStatsAllRepsFiltered$five_prime,
#                          POS_DIFF, title = "NHF1 CPD All Reps TGG-Filtered C>T Mismatch Pos Diff",
#                          xAxisBreaks = c(22:30))
# 
# plotGroupedPositionStats(arabidopsisCPDGroupedStatsAllRepsFiltered$three_prime,
#                          arabidopsisCPDGroupedStatsAllRepsFiltered$five_prime,
#                          POS_DIFF, title = "Arabidopsis CPD All Reps TGG-Filtered C>T Mismatch Pos Diff",
#                          xAxisBreaks = c(24:30))
# 
# plotGroupedPositionStats(yeastCPDGroupedStatsAllRepsFiltered$three_prime,
#                          yeastCPDGroupedStatsAllRepsFiltered$five_prime,
#                          POS_DIFF, title = "Yeast CPD All Reps TGG-Filtered C>T Mismatch Pos Diff",
#                          xAxisBreaks = c(22:25))
```


```{r read_figure_4_data}
NHF1ThreePrimeMismatchCPDNucFreqAllRepsFiltered = fread(generateFilePath(
  NHF1ExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
  strandPolarity = "three_prime", anchored = TRUE, filtering = "TGG_filtered"
))
NHF1ThreePrimeMMAnchoredCPDSeqFreqAllRepsFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1ThreePrimeMMAnchoredCPDSeqFreqAllRepsFiltered)) {
  NHF1ThreePrimeMMAnchoredCPDSeqFreqAllRepsFiltered[[i]] = fread(generateFilePath(
    NHF1ExcisionSiteDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    timepoint = names(NHF1ThreePrimeMMAnchoredCPDSeqFreqAllRepsFiltered)[i],
    strandPolarity = "three_prime", anchored = TRUE, sequence = "HYV", filtering = "TGG_filtered"
  ))
}

NHF1GeneralReadsCPDNucFreqAllReps1hFilteredCombined = fread(generateFilePath(
  NHF1GeneralReadSequenceDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "NHF1", lesion = "CPD", repitition = "all_reps", timepoint = "1h",
  filtering = "TGG_filtered", additionalInformation = "combined_read_lengths"
))
NHF1GeneralReadsCPDSeqFreqAllReps1hFilteredCombined = fread(generateFilePath(
  NHF1GeneralReadSequenceDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
  cellType = "NHF1", lesion = "CPD", timepoint = "1h", repitition = "all_reps",
  sequence = "HYV", filtering = "TGG_filtered", additionalInformation = "combined_read_lengths"
))
arabidopsisGeneralReadsCPDSeqFreqAllRepsZT2FilteredCombined = fread(generateFilePath(
  arabidopsisGeneralReadSequenceDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
  cellType = "Arabidopsis", lesion = "CPD", timepoint = "ZT2",
  repitition = "all_reps", sequence = "HYV", filtering = "TGG_filtered",
  additionalInformation = "combined_read_lengths"
))
yeastGeneralReadsCPDSeqFreqAllReps20MinFilteredCombined = fread(generateFilePath(
  yeastGeneralReadSequenceDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
  cellType = "yeast", lesion = "CPD", timepoint = "20min",
  repitition = "all_reps", sequence = "HYV", filtering = "TGG_filtered",
  additionalInformation = "combined_read_lengths"
))
EColiGeneralReadsCPDSeqFreqAllReps5MinFilteredCombined = fread(generateFilePath(
  EColiGeneralReadSequenceDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
  cellType = "E_coli_WT", lesion = "CPD", timepoint = "5min",
  repitition = "all_reps", sequence = "HYV", filtering = "TGG_filtered",
  additionalInformation = "combined_read_lengths"
))
```


```{r figure_4A, fig.width = 6.5, fig.height = 6}
plotNucFreqVsReadLengthBarPlot(NHF1ThreePrimeMismatchCPDNucFreqAllRepsFiltered[Timepoint == "1h"], 
                               title = "NHF1 CPD 1h All Reps TGG-Filtered 3' Anchored Nuc Freq",
                               secondaryYAxisLabel = "5' Cut Site Distance",
                               posType = FIVE_PRIME, showFivePrimeCutSite = TRUE,
                               expansionOffset = 10, minReadLength = 19, maxReadLength = 25,
                               xAxisLabel = "Mismatch Relative Position", positionOffset = 1)
```
```{r figure_4B, fig.width = 5, fig.height = 6}
# ### Just the 1h timepoint (asterisks) ###
# plotSequenceFrequencies(NHF1ThreePrimeMMAnchoredCPDSeqFreqAllRepsFiltered$`1h`,
#                         title = "NHF1 CPD 1h All Reps TGG-Filtered 3' Anchored HYV Freq",
#                         secondaryYAxisLabel = "5' Cut Site Distance", yAxisLabel = "YR Frequency",
#                         showFivePrimeCutSite = TRUE, querySequences = c("333"),
#                         expansionOffset = 10, minReadLength = 19, maxReadLength = 25, endBasedBackgroundNum = 6,
#                         endBasedCheckPositions = -9, xAxisLabel = "Mismatch Relative Position",
#                         significanceAsteriskBreakpoints = c(1e-3, 1e-10, 1e-20))

## Just the 1h timepoint (full p-values) ###
plotSequenceFrequencies(NHF1ThreePrimeMMAnchoredCPDSeqFreqAllRepsFiltered$`1h`,
                        title = "NHF1 CPD 1h All Reps TGG-Filtered 3' Anchored HYV Freq",
                        secondaryYAxisLabel = "5' Cut Site Distance", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"), xAxisOffset = 1,
                        expansionOffset = 10, minReadLength = 19, maxReadLength = 25, endBasedBackgroundNum = 6,
                        endBasedCheckPositions = -9, displayPValue = FALSE, xAxisLabel = "Mismatch Relative Position",
                        xAxisBreaks = c(-2,-10,-20,-30), xlim = c(-35, -4))

### All timepoints together ###
# plotSequenceFrequencies(NHF1ThreePrimeMMAnchoredCPDSeqFreqAllRepsFiltered,
#                         title = "NHF1 CPD All Reps TGG-Filtered 3' Anchored YR Freq",
#                         secondaryYAxisLabel = "5' Cut Site Distance", yAxisLabel = "YR Frequency",
#                         showFivePrimeCutSite = TRUE, querySequences = c("CG","CA","TG","TA"),
#                         expansionOffset = 10, minReadLength = 19, maxReadLength = 25, endBasedBackgroundNum = 8,
#                         endBasedCheckPositions = -10, displayPValue = TRUE, xAxisLabel = "Mismatch Relative Position")
```
```{r figure_4C, fig.width = 6.5, fig.height = 4.75}
plotNucFreqVsReadLengthBarPlot(NHF1GeneralReadsCPDNucFreqAllReps1hFilteredCombined[Position <= 7], 
                               title = "NHF1 CPD 1h All Reps TGG-Filtered Nuc Freq",
                               secondaryYAxisLabel = "Read Length", ylim = c(0,1),
                               posType = FIVE_PRIME, showFivePrimeCutSite = TRUE,
                               expansionOffset = 10, xAxisBreaks = c(-9,0,7),
                               xAxisLabel = "5' Relative Position")
```
```{r figure_4D, fig.width = 5, fig.height = 4}
plotSequenceFrequencies(NHF1GeneralReadsCPDSeqFreqAllReps1hFilteredCombined[Position <= 7],
                        title = "NHF1 CPD 1h All Reps TGG-Filtered General Reads HYV Freq",
                        secondaryYAxisLabel = "Read Length", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"), xAxisOffset = 1,
                        expansionOffset = 10, backgroundPositions = -9:-4,
                        checkPositions = -1, displayPValue = FALSE, xAxisLabel = "5' Relative Position",
                        xAxisBreaks = c(-8, 0, 8), ylimMax = 0.6, ylimMaxModifier = 1)
```
```{r figure_4E, fig.width = 5, fig.height = 4}
plotSequenceFrequencies(arabidopsisGeneralReadsCPDSeqFreqAllRepsZT2FilteredCombined[Position <= 7],
                        title = "Arabidopsis CPD ZT2 All Reps TGG-Filtered General Reads HYV Freq",
                        secondaryYAxisLabel = "Read Length", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"), xAxisOffset = 1,
                        expansionOffset = 10, backgroundPositions = -9:-4,
                        checkPositions = -1, displayPValue = FALSE, xAxisLabel = "5' Relative Position",
                        xAxisBreaks = c(-8, 0, 8), ylimMax = 0.6, ylimMaxModifier = 1)
```
```{r figure_4F, fig.width = 5, fig.height = 4}
plotSequenceFrequencies(yeastGeneralReadsCPDSeqFreqAllReps20MinFilteredCombined[Position <= 7],
                        title = "Yeast CPD 20min All Reps TGG-Filtered General Reads HYV Freq",
                        secondaryYAxisLabel = "Read Length", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"), xAxisOffset = 1,
                        expansionOffset = 10, backgroundPositions = -9:-4,
                        checkPositions = -1, displayPValue = FALSE, xAxisLabel = "5' Relative Position",
                        xAxisBreaks = c(-8, 0, 8), ylimMax = 0.6, ylimMaxModifier = 1)
```
```{r figure_4G, fig.width = 5, fig.height = 4}
plotSequenceFrequencies(EColiGeneralReadsCPDSeqFreqAllReps5MinFilteredCombined[Position <= 5],
                        title = "E. coli CPD 5min All Reps TGG-Filtered General Reads HYV Freq",
                        secondaryYAxisLabel = "Read Length", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"), xAxisOffset = 1,
                        expansionOffset = 10, xAxisLabel = "5' Relative Position",
                        xlim = c(-9,7), xAxisBreaks = c(-8, 0, 6), ylimMax = 0.6, ylimMaxModifier = 1)
```


```{r read_figure_5_data}
# Note: This file was created manually by combining/modifying a couple different files counting tandem mismatches.
NHF1TandemMismatchCountsComparisonAllReps = fread(generateFilePath(
  NHF1MismatchesByReadDataDirectory, dataTypeStrings$tandemMismatchCountsComparison, ".tsv",
  cellType = "NHF1", repitition = "all_reps"
))
```


```{r figure_5B, fig.width = 8, fig.height = 5}
NHF1TandemMismatchCountsComparisonAllReps = NHF1TandemMismatchCountsComparisonAllReps[
  Lesion == "CPD" & Timepoint == "1h"
]
print(
    ggplot(NHF1TandemMismatchCountsComparisonAllReps, aes(Mismatch_Type, Tandem_Mismatches_Count, fill = Search_Method)) +
      geom_bar(stat = "identity", position = "dodge") +
      labs(title = "NHF1 CPD 1h All Reps Tandem Mismatch Counts", x = "Mismatch Type", y = "Tandem Mismatch Count") +
      geom_text(aes(label=Tandem_Mismatches_Count), position = position_dodge(width = 0.9), vjust = -0.25, size = 5) +
      blankBackground + defaultTextScaling + scale_fill_brewer(palette = "Set1") +
      coord_cartesian(ylim = c(0,200000), expand = FALSE)
  )
```



```{r read_figure_6_data}
NHF1CPDCCtoTTMismatchDataAllRepsFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDCCtoTTMismatchDataAllRepsFiltered)) {
  NHF1CPDCCtoTTMismatchDataAllRepsFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    filtering = "TGG_filtered", includedMismatches = "CC>TT",
    timepoint = names(NHF1CPDCCtoTTMismatchDataAllRepsFiltered)[i]
  ))
}

NHF1CPDCCtoTTMMFreqZScoresAllRepsFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDCCtoTTMMFreqZScoresAllRepsFiltered)) {
  NHF1CPDCCtoTTMMFreqZScoresAllRepsFiltered[[i]] = fread(generateFilePath(
    directory = NHF1SequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    filtering = "TGG_filtered", includedMismatches = "CC>TT",
    timepoint = names(NHF1CPDCCtoTTMMFreqZScoresAllRepsFiltered)[i]
  ))
}

NHF1ThreePrimeMMAnchoredCPDCCtoTTSeqFreqAllReps1hFiltered = fread(generateFilePath(
    NHF1ExcisionSiteDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv", timepoint = "1h",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps", includedMismatches = "CC>TT",
    strandPolarity = "three_prime", anchored = TRUE, sequence = "HYV", filtering = "TGG_filtered",
))

NHF1CPDCCToTTGroupedStatsAllRepsFiltered = list(three_prime = NULL, five_prime = NULL)
for (i in seq_along(NHF1CPDCCToTTGroupedStatsAllRepsFiltered)) {
  NHF1CPDCCToTTGroupedStatsAllRepsFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    strandPolarity = names(NHF1CPDCCToTTGroupedStatsAllRepsFiltered)[i], includedMismatches = "CC>TT"
  ))
}

arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered = list(three_prime = NULL, five_prime = NULL)
for (i in seq_along(arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered)) {
  arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered[[i]] = fread(generateFilePath(
    arabidopsisSequenceAndPositionDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
    cellType = "Arabidopsis", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    strandPolarity = names(arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered)[i], includedMismatches = "CC>TT"
  ))
}

yeastCPDCCToTTGroupedStatsAllRepsFiltered = list(three_prime = NULL, five_prime = NULL)
for (i in seq_along(yeastCPDCCToTTGroupedStatsAllRepsFiltered)) {
  yeastCPDCCToTTGroupedStatsAllRepsFiltered[[i]] = fread(generateFilePath(
    yeastSequenceAndPositionDataDirectory, dataTypeStrings$mismatchPositionStats, ".tsv",
    cellType = "yeast", lesion = "CPD", repitition = "all_reps", filtering = "TGG_filtered",
    strandPolarity = names(yeastCPDCCToTTGroupedStatsAllRepsFiltered)[i], includedMismatches = "CC>TT"
  ))
}
```


```{r figure_6A, fig.width = 5, fig.height = 5.5}
### Just the 1h timepoint ###
plotPositionAcrossTimepointAndReadLength(NHF1CPDCCtoTTMismatchDataAllRepsFiltered["1h"],
                                         title = "NHF1 CPD 1h All Reps TGG-Filtered CC>TT Mismatch Pos Freq",
                                         zScoreTables = NHF1CPDCCtoTTMMFreqZScoresAllRepsFiltered,
                                         yAxisLabel = "Relative CC>TT Frequency", xAxisBreaks = c(-30,-20,-10,-1))

### All timepoints together ###
# plotPositionAcrossTimepointAndReadLength(NHF1CPDCCtoTTMismatchDataAllRepsFiltered,
#                                          title = "NHF1 CPD All Reps TGG-Filtered CC>TT Mismatch Pos Freq",
#                                          zScoreTables = NHF1CPDCCtoTTMMFreqZScoresAllRepsFiltered,
#                                          yAxisLabel = "Relative CC>TT Frequency")
```
```{r figure_6B, fig.width = 5, fig.height = 5}
plotSequenceFrequencies(NHF1ThreePrimeMMAnchoredCPDCCtoTTSeqFreqAllReps1hFiltered,
                        title = "NHF1 CPD 1h All Reps TGG-Filtered CC>TT 3' Anchored HYV Freq",
                        secondaryYAxisLabel = "5' Cut Site Distance", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"),
                        expansionOffset = 10, minReadLength = 19, maxReadLength = 25, endBasedBackgroundNum = 6,
                        endBasedCheckPositions = -9, displayPValue = FALSE, xAxisLabel = "Mismatch Relative Position",
                        xAxisOffset = 1, xAxisBreaks = c(-2,-10,-20,-30), xlim = c(-35,-4))
```
```{r figure_6C, fig.width = 7.5, fig.height = 5}
### Just the 1h timepoint ###
plotGroupedPositionStats(NHF1CPDCCToTTGroupedStatsAllRepsFiltered$three_prime[Timepoint == "1h"],
                         NHF1CPDCCToTTGroupedStatsAllRepsFiltered$five_prime[Timepoint == "1h"],
                         POS_DIFF, title = "NHF1 CPD 1h All Reps TGG-Filtered CC>TT Mismatch Pos Diff",
                         xAxisBreaks = c(22:30))

### All timepoints together ###
# plotGroupedPositionStats(NHF1CPDCCToTTGroupedStatsAllRepsFiltered$three_prime,
#                          NHF1CPDCCToTTGroupedStatsAllRepsFiltered$five_prime,
#                          POS_DIFF, title = "NHF1 CPD All Reps TGG-Filtered CC>TT Mismatch Pos Diff",
#                          xAxisBreaks = c(22:30))
```
```{r figure_6D, fig.width = 7.5, fig.height = 5}
### Just the 1h timepoint ###
plotGroupedPositionStats(arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered$three_prime[Timepoint == "ZT2"],
                         arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered$five_prime[Timepoint == "ZT2"],
                         POS_DIFF, title = "Arabidopsis CPD 1h All Reps TGG-Filtered CC>TT Mismatch Pos Diff",
                         xAxisBreaks = c(24:30))

### All timepoints together ###
# plotGroupedPositionStats(arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered$three_prime,
#                          arabidopsisCPDCCToTTGroupedStatsAllRepsFiltered$five_prime,
#                          POS_DIFF, title = "Arabidopsis CPD All Reps TGG-Filtered CC>TT Mismatch Pos Diff",
#                          xAxisBreaks = c(24:30))
```
```{r figure_6E, fig.width = 7.5, fig.height = 5}
### Just the 1h timepoint ###
plotGroupedPositionStats(yeastCPDCCToTTGroupedStatsAllRepsFiltered$three_prime[Timepoint == "20min"],
                         yeastCPDCCToTTGroupedStatsAllRepsFiltered$five_prime[Timepoint == "20min"],
                         POS_DIFF, title = "Yeast CPD 1h All Reps TGG-Filtered CC>TT Mismatch Pos Diff",
                         xAxisBreaks = c(22:25))

### All timepoints together ###
# plotGroupedPositionStats(yeastCPDCCToTTGroupedStatsAllRepsFiltered$three_prime,
#                          yeastCPDCCToTTGroupedStatsAllRepsFiltered$five_prime,
#                          POS_DIFF, title = "Yeast CPD All Reps TGG-Filtered CC>TT Mismatch Pos Diff",
#                          xAxisBreaks = c(22:25))
```


```{r figure_S1}
plotMismatchPositionFrequencies(NHF1CPDMismatchDataAllReps$`1h`, omittedTypes = c("C>T"),
                                title = "NHF1 CPD 1h All Reps !C>T Mismatch Pos Frequencies",
                                xAxisBreaks = c(-1,-10,-20,-30))
```

```{r read_figure_S2_data}
NHF1SixFourMismatchDataAllReps = NHF1SixFourTimepoints
for (i in seq_along(NHF1SixFourMismatchDataAllReps)) {
  NHF1SixFourMismatchDataAllReps[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "NHF1", lesion = "6-4", repitition = "all_reps",
    timepoint = names(NHF1SixFourMismatchDataAllReps)[i]
  ))
}

GM12878CisplatinMismatchDataAllReps = GM12878CisplatinTimepoints
for (i in seq_along(GM12878CisplatinMismatchDataAllReps)) {
  GM12878CisplatinMismatchDataAllReps[[i]] = fread(generateFilePath(
    GM12878SequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "GM12878", lesion = "cisplatin", repitition = "all_reps",
    timepoint = names(GM12878CisplatinMismatchDataAllReps)[i]
  ))
}
```


```{r figure_S2}
plotMismatchTypeFrequencies(NHF1SixFourMismatchDataAllReps$`20min`,
                            title = "NHF1 6-4 20min All Reps Mismatch Type Frequencies")
plotMismatchTypeFrequencies(GM12878CisplatinMismatchDataAllReps$`90min`,
                            title = "GM12878 Cisplatin 90min All Reps Mismatch Type Frequencies")


print("NHF1 6-4 20min all reps C>T mismatch dipyrimidine context frequency:")
plotTrinucleotideContext(NHF1SixFourMismatchDataAllReps$`20min`, includedTypes = c("C>T"),
                         title = "NHF1 6-4 20min All Reps Mismatch Trinuc Context",
                         printDipyFreq = TRUE)

print("GM12878 cisplatin 90min all reps C>T mismatch dipyrimidine context frequency:")
plotTrinucleotideContext(GM12878CisplatinMismatchDataAllReps$`90min`, includedTypes = c("C>T"),
                         title = "GM12878 Cisplatin 90min All Reps Mismatch Trinuc Context",
                         printDipyFreq = TRUE)


plotMismatchPositionFrequencies(NHF1SixFourMismatchDataAllReps$`20min`, includedTypes = c("C>T"),
                                title = "NHF1 6-4 20min All Reps Mismatch Pos Frequencies",
                                xAxisBreaks = c(-1,-10,-20,-30))

plotMismatchPositionFrequencies(GM12878CisplatinMismatchDataAllReps$`90min`, includedTypes = c("C>T"),
                                title = "GM12878 Cisplatin 90min All Reps Mismatch Pos Frequencies",
                                xAxisBreaks = c(-1,-10,-20,-30))
```


```{r read_figure_S3_data}
arabidopsisCPDMismatchDataAllReps = arabidopsisCPDTimepoints
for (i in seq_along(arabidopsisCPDMismatchDataAllReps)) {
  arabidopsisCPDMismatchDataAllReps[[i]] = fread(generateFilePath(
    arabidopsisSequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "Arabidopsis", lesion = "CPD", repitition = "all_reps",
    timepoint = names(arabidopsisCPDMismatchDataAllReps)[i]
  ))
}

yeastCPDMismatchDataAllReps = yeastCPDTimepoints
for (i in seq_along(yeastCPDMismatchDataAllReps)) {
  yeastCPDMismatchDataAllReps[[i]] = fread(generateFilePath(
    yeastSequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "yeast", lesion = "CPD", repitition = "all_reps",
    timepoint = names(yeastCPDMismatchDataAllReps)[i]
  ))
}
```


```{r figure_S3}
plotMismatchTypeFrequencies(arabidopsisCPDMismatchDataAllReps$ZT2,
                            title = "Arabidopsis CPD ZT2 All Reps Mismatch Type Frequencies")

plotMismatchTypeFrequencies(yeastCPDMismatchDataAllReps$`20min`,
                            title = "Yeast CPD 20min All Reps Mismatch Type Frequencies")


print("Arabidopsis CPD 30min all reps C>T mismatch dipyrimidine context frequency:")
plotTrinucleotideContext(arabidopsisCPDMismatchDataAllReps$ZT2, includedTypes = c("C>T"),
                         title = "Arabidopsis CPD ZT2 All Reps Mismatch Trinuc Context",
                         printDipyFreq = TRUE)

print("Yeast CPD 20min all reps C>T mismatch dipyrimidine context frequency:")
plotTrinucleotideContext(yeastCPDMismatchDataAllReps$`20min`, includedTypes = c("C>T"),
                         title = "Yeast CPD 20min All Reps Mismatch Trinuc Context",
                         printDipyFreq = TRUE)


plotMismatchPositionFrequencies(arabidopsisCPDMismatchDataAllReps$ZT2, includedTypes = c("C>T"),
                                title = "Arabidopsis CPD ZT2 All Reps Mismatch Pos Frequencies",
                                xAxisBreaks = c(-1,-10,-20,-30))

plotMismatchPositionFrequencies(yeastCPDMismatchDataAllReps$`20min`, includedTypes = c("C>T"),
                                title = "Yeast CPD 20min All Reps Mismatch Pos Frequencies",
                                xAxisBreaks = c(-1,-10,-20,-30))
```


```{r read_figure_S4_data}
CSBCPDMismatchDataAllReps = CSBCPDTimepoints
for (i in seq_along(CSBCPDMismatchDataAllReps)) {
  CSBCPDMismatchDataAllReps[[i]] = fread(generateFilePath(
    CSBSequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "CS-B", lesion = "CPD", repitition = "all_reps",
    timepoint = names(CSBCPDMismatchDataAllReps)[i]
  ))
}

XPCCPDMismatchDataAllReps = XPCCPDTimepoints
for (i in seq_along(XPCCPDMismatchDataAllReps)) {
  XPCCPDMismatchDataAllReps[[i]] = fread(generateFilePath(
    XPCSequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "XP-C", lesion = "CPD", repitition = "all_reps",
    timepoint = names(XPCCPDMismatchDataAllReps)[i]
  ))
}
```


```{r figure_S4}
plotMismatchTypeFrequencies(CSBCPDMismatchDataAllReps$`1h`,
                            title = "CS-B CPD 1h All Reps Mismatch Type Frequencies")

plotMismatchTypeFrequencies(XPCCPDMismatchDataAllReps$`1h`,
                            title = "XP-C CPD 1h All Reps Mismatch Type Frequencies")


print("CS-B CPD 1h all reps C>T mismatch dipyrimidine context frequency:")
plotTrinucleotideContext(CSBCPDMismatchDataAllReps$`1h`, includedTypes = c("C>T"),
                         title = "CS-B CPD 1h All Reps Mismatch Trinuc Context",
                         printDipyFreq = TRUE)

print("XPC CPD 1h all reps C>T mismatch dipyrimidine context frequency:")
plotTrinucleotideContext(XPCCPDMismatchDataAllReps$`1h`, includedTypes = c("C>T"),
                         title = "XP-C CPD 1h All Reps Mismatch Trinuc Context",
                         printDipyFreq = TRUE)


plotMismatchPositionFrequencies(CSBCPDMismatchDataAllReps$`1h`, includedTypes = c("C>T"),
                                title = "CS-B CPD 1h All Reps Mismatch Pos Frequencies",
                                ylim = c(0,0.25), xAxisBreaks = c(-1,-10,-20,-30))

plotMismatchPositionFrequencies(XPCCPDMismatchDataAllReps$`1h`, includedTypes = c("C>T"),
                                title = "XP-C CPD 1h All Reps Mismatch Pos Frequencies",
                                ylim = c(0,0.25), xAxisBreaks = c(-1,-10,-20,-30))
```


```{r read_figure_S5_data}
NHF1FivePrimeMismatchCPDNucFreqAllRepsTGGFiltered = fread(generateFilePath(
  NHF1ExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
  strandPolarity = "five_prime", anchored = TRUE, filtering = "TGG_filtered"
))
NHF1FivePrimeMismatchCPDNucFreqAllRepsTGFiltered = fread(generateFilePath(
  NHF1ExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
  strandPolarity = "five_prime", anchored = TRUE, filtering = "TG_filtered"
))

NHF1CPDMismatchDataAllRepsTGFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDMismatchDataAllRepsTGFiltered)) {
  NHF1CPDMismatchDataAllRepsTGFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "NHF1", lesion = "CPD", timepoint = names(NHF1CPDMismatchDataAllRepsTGFiltered)[i],
    repitition = "all_reps", filtering = "TG_filtered"
  ))
}

NHF1CPDMMFreqZScoresAllRepsTGFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDMMFreqZScoresAllRepsTGFiltered)) {
  NHF1CPDMMFreqZScoresAllRepsTGFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps", filtering = "TG_filtered",
    timepoint = names(NHF1CPDMMFreqZScoresAllRepsTGFiltered)[i], includedMismatches = c("C>T")
  ))
}
```


```{r figure_S5A, fig.width = 6.5, fig.height = 6}
### Just the 1h timepoint ###
plotNucFreqVsReadLengthBarPlot(NHF1FivePrimeMismatchCPDNucFreqAllRepsTGGFiltered[Timepoint == "1h"], 
                               title = "NHF1 CPD 1h All Reps TGG-Filtered 5' Anchored Nuc Freq",
                               secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE,
                               expansionOffset = 10, minReadLength = 1, maxReadLength = 8,
                               xAxisLabel = "Mismatch Relative Position", positionOffset = -1)

### All timepoints together ###
# plotNucFreqVsReadLengthBarPlot(NHF1FivePrimeMismatchCPDNucFreqAllRepsTGGFiltered, 
#                                title = "NHF1 CPD All Reps TGG-Filtered 5' Anchored Nuc Freq",
#                                secondaryYAxisLabel = "3' Cut Site Distance",
#                                posType = FIVE_PRIME, showThreePrimeCutSite = TRUE,
#                                expansionOffset = 10, minReadLength = 1, maxReadLength = 8,
#                                xAxisLabel = "Mismatch Relative Position")
```
```{r figure_S5B, fig.width = 6.5, fig.height = 6}
### Just the 1h timepoint ###
plotNucFreqVsReadLengthBarPlot(NHF1FivePrimeMismatchCPDNucFreqAllRepsTGFiltered[Timepoint == "1h"], 
                               title = "NHF1 CPD 1h All Reps TG-Filtered 5' Anchored Nuc Freq",
                               secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = FIVE_PRIME, showThreePrimeCutSite = TRUE,
                               expansionOffset = 10, minReadLength = 1, maxReadLength = 8,
                               xAxisLabel = "Mismatch Relative Position", positionOffset = -1)

### All timepoints together ###
# plotNucFreqVsReadLengthBarPlot(NHF1FivePrimeMismatchCPDNucFreqAllRepsTGFiltered, 
#                                title = "NHF1 CPD All Reps TG-Filtered 5' Anchored Nuc Freq",
#                                secondaryYAxisLabel = "3' Cut Site Distance",
#                                posType = FIVE_PRIME, showThreePrimeCutSite = TRUE,
#                                expansionOffset = 10, minReadLength = 1, maxReadLength = 8,
#                                xAxisLabel = "Mismatch Relative Position")
```
```{r figure_S5C, fig.width = 5, fig.height = 5}
### Just the 1h timepoint ###
plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGGFiltered["1h"], includedTypes = c("C>T"),
                                         title = "NHF1 CPD 1h All Reps TG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGFiltered,
                                         posType = THREE_PRIME, yAxisLabel = "Relative C>T Frequency",
                                         xAxisBreaks = c(-1,-10,-20,-30))


### All timepoints together ###
# plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGFiltered, includedTypes = c("C>T"),
#                                          title = "NHF1 CPD All Reps TG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGFiltered,
#                                          posType = THREE_PRIME, yAxisLabel = "Relative C>T Frequency")
```
```{r figure_S5D, fig.width = 5, fig.height = 5}
### Just the 1h timepoint ###
plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGFiltered["1h"], includedTypes = c("C>T"),
                                         title = "NHF1 CPD 1h All Reps TG-Filtered C>T Mismatch Pos Freq",
                                         zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGFiltered,
                                         posType = THREE_PRIME, yAxisLabel = "Relative C>T Frequency",
                                         xAxisBreaks = c(-1,-10,-20,-30))


### All timepoints together ###
# plotPositionAcrossTimepointAndReadLength(NHF1CPDMismatchDataAllRepsTGFiltered, includedTypes = c("C>T"),
#                                          title = "NHF1 CPD All Reps TG-Filtered C>T Mismatch Pos Freq",
#                                          zScoreTables = NHF1CPDMMFreqZScoresAllRepsTGFiltered,
#                                          posType = THREE_PRIME, yAxisLabel = "Relative C>T Frequency")
```


```{r read_figure_S6_data}
NHF1GeneralReadsSixFourSeqFreqAllReps20minFilteredCombined = fread(generateFilePath(
  NHF1GeneralReadSequenceDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
  cellType = "NHF1", lesion = "6-4", timepoint = "20min",
  repitition = "all_reps", sequence = "HYV", filtering = "TGG_filtered",
  additionalInformation = "combined_read_lengths"
))
GM12878GeneralReadsCisplatinSeqFreqRep1_90minFilteredCombined = fread(generateFilePath(
  GM12878GeneralReadSequenceDataDirectory, dataTypeStrings$sequenceFrequency, ".tsv",
  cellType = "GM12878", lesion = "cisplatin", timepoint = "90min",
  repitition = "rep1", sequence = "HYV", filtering = "TGG_filtered",
  additionalInformation = "combined_read_lengths"
))
```


```{r figure_S6, fig.width = 5, fig.height = 4}
plotSequenceFrequencies(NHF1GeneralReadsSixFourSeqFreqAllReps20minFilteredCombined[Position <= 7],
                        title = "NHF1 6-4 20min All Reps TGG-Filtered General Reads HYV Freq",
                        secondaryYAxisLabel = "Read Length", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"), xAxisBreaks = c(-8, 0, 8),
                        expansionOffset = 10, minReadLength = 23, maxReadLength = 29, backgroundPositions = -9:-4,
                        checkPositions = -1, displayPValue = FALSE, xAxisLabel = "5' Relative Position",
                        xAxisOffset = 1, ylimMax = 0.6, ylimMaxModifier = 1)
plotSequenceFrequencies(GM12878GeneralReadsCisplatinSeqFreqRep1_90minFilteredCombined[Position <= 7],
                        title = "GM12878 cisplatin 90min Rep1 TGG-Filtered General Reads HYV Freq",
                        secondaryYAxisLabel = "Read Length", yAxisLabel = "HYV Frequency",
                        showFivePrimeCutSite = TRUE, querySequences = c("333"), xAxisBreaks = c(-8, 0, 8),
                        expansionOffset = 10, minReadLength = 23, maxReadLength = 29, backgroundPositions = -9:-4,
                        checkPositions = -1, displayPValue = FALSE, xAxisLabel = "5' Relative Position",
                        xAxisOffset = 1, ylimMax = 0.6, ylimMaxModifier = 1)
```


```{r read_figure_S7_data}
NHF1FivePrimeMismatchCPDNucFreqAllRepsFiltered = fread(generateFilePath(
  NHF1ExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
  strandPolarity = "five_prime", anchored = TRUE, filtering = "TGG_filtered"
))
arabidopsisFivePrimeMismatchCPDNucFreqAllRepsFiltered = fread(generateFilePath(
  arabidopsisExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "Arabidopsis", lesion = "CPD", repitition = "all_reps",
  strandPolarity = "five_prime", anchored = TRUE, filtering = "TGG_filtered"
))
yeastFivePrimeMismatchCPDNucFreqAllRepsFiltered = fread(generateFilePath(
  yeastExcisionSiteDataDirectory, dataTypeStrings$nucleotideFrequencyData, ".tsv",
  cellType = "yeast", lesion = "CPD", repitition = "all_reps",
  strandPolarity = "five_prime", anchored = TRUE, filtering = "TGG_filtered"
))
```


```{r figure_S7, fig.width = 6.5, fig.height = 6}
plotNucFreqVsReadLengthBarPlot(NHF1FivePrimeMismatchCPDNucFreqAllRepsFiltered[Timepoint == "1h"], 
                               title = "NHF1 CPD 1h All Reps TGG-Filtered 5' Anchored Nuc Freq",
                               secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = THREE_PRIME, showThreePrimeCutSite = TRUE,
                               expansionOffset = 10, minReadLength = 6, maxReadLength = 12,
                               xAxisLabel = "Mismatch Relative Position", positionOffset = -1)
plotNucFreqVsReadLengthBarPlot(arabidopsisFivePrimeMismatchCPDNucFreqAllRepsFiltered[Timepoint == "ZT2"], 
                               title = "Arabidopsis CPD 1h All Reps TGG-Filtered 5' Anchored Nuc Freq",
                               secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = THREE_PRIME, showThreePrimeCutSite = TRUE,
                               expansionOffset = 10, minReadLength = 6, maxReadLength = 12,
                               xAxisLabel = "Mismatch Relative Position", positionOffset = -1)
plotNucFreqVsReadLengthBarPlot(yeastFivePrimeMismatchCPDNucFreqAllRepsFiltered[Timepoint == "20min"], 
                               title = "Yeast CPD 1h All Reps TGG-Filtered 5' Anchored Nuc Freq",
                               secondaryYAxisLabel = "3' Cut Site Distance",
                               posType = THREE_PRIME, showThreePrimeCutSite = TRUE,
                               expansionOffset = 10, minReadLength = 6, maxReadLength = 12,
                               xAxisLabel = "Mismatch Relative Position", positionOffset = -1)
```



```{r read_figure_S8_data}
NHF1CPDGGtoAAMismatchDataAllRepsFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDGGtoAAMismatchDataAllRepsFiltered)) {
  NHF1CPDGGtoAAMismatchDataAllRepsFiltered[[i]] = fread(generateFilePath(
    NHF1SequenceAndPositionDataDirectory, dataTypeStrings$simplifiedMismatches, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    filtering = "TGG_filtered", includedMismatches = "GG>AA",
    timepoint = names(NHF1CPDGGtoAAMismatchDataAllRepsFiltered)[i]
  ))
}

NHF1CPDGGtoAAMMFreqZScoresAllRepsFiltered = NHF1CPDTimepoints
for (i in seq_along(NHF1CPDGGtoAAMMFreqZScoresAllRepsFiltered)) {
  NHF1CPDGGtoAAMMFreqZScoresAllRepsFiltered[[i]] = fread(generateFilePath(
    directory = NHF1SequenceAndPositionDataDirectory, dataTypeStrings$mismatchFrequencyZScores, ".tsv",
    cellType = "NHF1", lesion = "CPD", repitition = "all_reps",
    filtering = "TGG_filtered", includedMismatches = "GG>AA",
    timepoint = names(NHF1CPDGGtoAAMMFreqZScoresAllRepsFiltered)[i]
  ))
}
```


```{r figure_S8, fig.width = 5, fig.height = 5.5}
plotPositionAcrossTimepointAndReadLength(NHF1CPDGGtoAAMismatchDataAllRepsFiltered["1h"],
                                         title = "NHF1 CPD All Reps GG>AA Mismatch Pos Freq",
                                         zScoreTables = NHF1CPDGGtoAAMMFreqZScoresAllRepsFiltered,
                                         yAxisLabel = "Relative GG>AA Frequency",
                                         xAxisBreaks = c(-1,-10,-20,-30))
```
